From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ProdPreva1l <prod.preva1l@gmail.com>
Date: Wed, 5 Jun 2024 20:44:50 +1000
Subject: [PATCH] Add various API for Folia plugin compatibility


diff --git a/src/main/java/io/papermc/paper/threadedregions/RegionizedServerInitEvent.java b/src/main/java/io/papermc/paper/threadedregions/RegionizedServerInitEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..d991df9523676ab7ba90f45ecae3ea9dac5b55d6
--- /dev/null
+++ b/src/main/java/io/papermc/paper/threadedregions/RegionizedServerInitEvent.java
@@ -0,0 +1,24 @@
+package io.papermc.paper.threadedregions;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+public class RegionizedServerInitEvent extends Event {
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+
+    @ApiStatus.Internal
+    public RegionizedServerInitEvent() {
+        super();
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 4a870c63cf240c4fa89534bf3ec1b06393efb967..6c3846695dfb84f7c2659dc2bc85249fabd82274 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2905,6 +2905,14 @@ public final class Bukkit {
         return server.isOwnedByCurrentRegion(entity);
     }
     // Paper end - Folia region threading API
+    // Confetti start - region threading API
+    /**
+     * Returns whether the current thread is ticking the global region.
+     */
+    public static boolean isGlobalTickThread() {
+        return server.isGlobalTickThread();
+    }
+    // Confetti end - region threading API
 
     @NotNull
     public static Server.Spigot spigot() {
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1d594c256200d46882fafce5d2b33e60f0fa3a59..6773c708593399745688b5f415b138d55c3a6028 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2555,6 +2555,12 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     boolean isOwnedByCurrentRegion(@NotNull Entity entity);
     // Paper end - Folia region threading API
+    // Confetti start - region threading API
+    /**
+     * Returns whether the current thread is ticking the global region.
+     */
+    public boolean isGlobalTickThread();
+    // Confetti end - region threading API
 
     // Purpur start
     /**
