From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 4 Mar 2022 20:33:33 +1000
Subject: [PATCH] Confetti config file


diff --git a/src/main/java/com/vulpeslab/confetti/config/ConfettiConfiguration.java b/src/main/java/com/vulpeslab/confetti/config/ConfettiConfiguration.java
new file mode 100644
index 0000000000000000000000000000000000000000..d0f12ab39527194bd9b47a7f04f2d3746a656ac5
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/config/ConfettiConfiguration.java
@@ -0,0 +1,64 @@
+package com.vulpeslab.confetti.config;
+
+import io.papermc.paper.configuration.ConfigurationPart;
+import org.spongepowered.configurate.objectmapping.meta.Comment;
+
+import java.util.List;
+
+@SuppressWarnings({ "InnerClassMayBeStatic" })
+public class ConfettiConfiguration extends ConfigurationPart {
+
+    public static final String HEADER = """
+            This is the main configuration file for Confetti.
+            There's quite alot to configure. Read the docs for more information.
+
+            Docs: https://github.com/vulpeslab/Confetti/blob/main/CONFETTI_YAML.md\s
+            """;
+
+    private static ConfettiConfiguration instance;
+
+    public static ConfettiConfiguration get() {
+        return instance;
+    }
+
+    static void set(ConfettiConfiguration instance) {
+        ConfettiConfiguration.instance = instance;
+    }
+
+    public Multithreading multithreading;
+
+    public class Multithreading extends ConfigurationPart {
+
+        public int threadCount = -1;
+        public int regionSize = 8;
+        public boolean runUnsupportedPluginsInSync = true;
+        public boolean allowUnsupportedPluginsToModifyChunksViaGlobalScheduler = true;
+
+    }
+
+    public Optimizations optimizations;
+
+    public class Optimizations extends ConfigurationPart {
+
+        public int entityActivationCheckFrequency = 20;
+        public boolean disableVanishApi = false;
+        public boolean useLazyExecuteWhenNotFlushing = true;
+        public boolean processTrackQueueInParallel = true;
+        public boolean flushQueueInParallel = true;
+        public int maximumTrackersPerEntity = 500;
+        public long trackerFullUpdateFrequency = 20;
+        public boolean writePlayerSavesAsync = true;
+        public ChunkPacketCaching chunkPacketCaching;
+
+        public class ChunkPacketCaching extends ConfigurationPart {
+
+            public boolean enabled = true;
+            public boolean useSoftReferences = true;
+            public long expireAfter = 1200;
+
+        }
+
+    }
+
+
+}
diff --git a/src/main/java/com/vulpeslab/confetti/config/ConfettiConfigurationLoader.java b/src/main/java/com/vulpeslab/confetti/config/ConfettiConfigurationLoader.java
new file mode 100644
index 0000000000000000000000000000000000000000..c51d25ddc99fa9d9d380105f063c47cd0438ef6a
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/config/ConfettiConfigurationLoader.java
@@ -0,0 +1,128 @@
+package com.vulpeslab.confetti.config;
+
+import io.papermc.paper.configuration.ConfigurationLoaders;
+import io.papermc.paper.configuration.ConfigurationPart;
+import io.papermc.paper.configuration.constraint.Constraint;
+import io.papermc.paper.configuration.constraint.Constraints;
+import io.papermc.paper.configuration.mapping.InnerClassFieldDiscoverer;
+import org.spongepowered.configurate.BasicConfigurationNode;
+import org.spongepowered.configurate.CommentedConfigurationNode;
+import org.spongepowered.configurate.ConfigurateException;
+import org.spongepowered.configurate.ConfigurationNode;
+import org.spongepowered.configurate.ConfigurationOptions;
+import org.spongepowered.configurate.objectmapping.ObjectMapper;
+import org.spongepowered.configurate.serialize.SerializationException;
+import org.spongepowered.configurate.yaml.YamlConfigurationLoader;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.function.UnaryOperator;
+
+import static io.leangen.geantyref.GenericTypeReflector.erase;
+
+public class ConfettiConfigurationLoader {
+
+    public static void init(File file) {
+        try {
+            YamlConfigurationLoader.Builder loaderBuilder = ConfigurationLoaders.naturallySorted();
+            loaderBuilder.defaultOptions(options -> options.header(ConfettiConfiguration.HEADER));
+
+            Path configFile = file.toPath();
+            YamlConfigurationLoader loader = loaderBuilder
+                    .defaultOptions(applyObjectMapperFactory(createObjectMapper().build()))
+                    .path(configFile)
+                    .build();
+            ConfigurationNode node;
+            if (Files.exists(configFile)) {
+                node = loader.load();
+            } else {
+                node = CommentedConfigurationNode.root(loader.defaultOptions());
+            }
+
+            String before = node.toString();
+            setFromProperties(node);
+
+            ConfettiConfiguration instance = node.require(ConfettiConfiguration.class);
+            transformLegacyConfig(node, instance);
+
+            for (Object key : node.childrenMap().keySet()) {
+                node.removeChild(key);
+            }
+
+            node.set(instance);
+
+            if (!node.toString().equals(before)) {
+                loader.save(node);
+            }
+
+            ConfettiConfiguration.set(instance);
+        } catch (ConfigurateException e) {
+            throw new RuntimeException("Could not load confetti.yml", e);
+        }
+    }
+
+    private static void setFromProperties(ConfigurationNode node) {
+        for (Map.Entry<Object, Object> property : System.getProperties().entrySet()) {
+            if (property.getKey().toString().startsWith("confetti.")) {
+                String key = property.getKey().toString().substring("confetti.".length());
+                try {
+                    node.node((Object[]) key.split("\\.")).set(property.getValue());
+                } catch (SerializationException e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+    }
+
+    private static ObjectMapper.Factory.Builder createObjectMapper() {
+        return ObjectMapper.factoryBuilder()
+                .addConstraint(Constraint.class, new Constraint.Factory())
+                .addConstraint(Constraints.Min.class, Number.class, new Constraints.Min.Factory())
+                .addDiscoverer(InnerClassFieldDiscoverer.globalConfig());
+    }
+
+    private static UnaryOperator<ConfigurationOptions> applyObjectMapperFactory(final ObjectMapper.Factory factory) {
+        return options -> options.serializers(builder -> builder
+                .register(type -> ConfigurationPart.class.isAssignableFrom(erase(type)), factory.asTypeSerializer())
+                .registerAnnotatedObjects(factory));
+    }
+
+    private static void transformLegacyConfig(ConfigurationNode node, ConfettiConfiguration config) {
+        // getAndRemove(node, "oldValue.oldie", value -> config.newSection.newValue.newie = value.getString());
+    }
+
+    private static void getAndRemove(ConfigurationNode node, String key, ExceptionableConsumer<ConfigurationNode> consumer) {
+        if (System.getProperty(key) != null) {
+            try {
+                BasicConfigurationNode systemNode = BasicConfigurationNode.root();
+                systemNode.set(System.getProperty(key));
+                consumer.accept(systemNode);
+                return;
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+
+        String[] parts = key.split("\\.");
+
+        for (int i = 0; i < parts.length; i++) {
+            if (node.isMap() && node.childrenMap().containsKey(parts[i])) {
+                if (i == parts.length - 1) {
+                    try {
+                        consumer.accept(node.childrenMap().get(parts[i]));
+                        node.removeChild(parts[i]);
+                    } catch (Exception e) {
+                        e.printStackTrace();
+                    }
+                } else {
+                    node = node.childrenMap().get(parts[i]);
+                }
+            } else {
+                return;
+            }
+        }
+    }
+
+}
diff --git a/src/main/java/com/vulpeslab/confetti/config/ExceptionableConsumer.java b/src/main/java/com/vulpeslab/confetti/config/ExceptionableConsumer.java
new file mode 100644
index 0000000000000000000000000000000000000000..4ba4ef9ddfc07f083f30db6712dcf798518a42d6
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/config/ExceptionableConsumer.java
@@ -0,0 +1,8 @@
+package com.vulpeslab.confetti.config;
+
+@FunctionalInterface
+public interface ExceptionableConsumer<T> {
+
+    void accept(T t) throws Exception;
+
+}
diff --git a/src/main/java/com/vulpeslab/confetti/permissions/ConfettiCommandPermissions.java b/src/main/java/com/vulpeslab/confetti/permissions/ConfettiCommandPermissions.java
new file mode 100644
index 0000000000000000000000000000000000000000..7ab9920729daf2cc82909d539c573028503b23ea
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/permissions/ConfettiCommandPermissions.java
@@ -0,0 +1,17 @@
+package com.vulpeslab.confetti.permissions;
+
+import org.bukkit.permissions.Permission;
+import org.bukkit.util.permissions.DefaultPermissions;
+import org.jetbrains.annotations.NotNull;
+
+public class ConfettiCommandPermissions {
+    private static final String ROOT = "confetti.command";
+    private static final String PREFIX = ROOT + ".";
+
+    public static void registerPermissions(@NotNull Permission parent) {
+        Permission commands = DefaultPermissions.registerPermission(ROOT, "Gives the user the ability to use all Confetti commands", parent);
+
+
+        commands.recalculatePermissibles();
+    }
+}
diff --git a/src/main/java/com/vulpeslab/confetti/permissions/ConfettiPermissions.java b/src/main/java/com/vulpeslab/confetti/permissions/ConfettiPermissions.java
new file mode 100644
index 0000000000000000000000000000000000000000..e2101157924ebf14edd77927b8563e273be37ce0
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/permissions/ConfettiPermissions.java
@@ -0,0 +1,19 @@
+package com.vulpeslab.confetti.permissions;
+
+import org.bukkit.permissions.Permission;
+import org.bukkit.permissions.PermissionDefault;
+import org.bukkit.util.permissions.DefaultPermissions;
+
+public class ConfettiPermissions {
+    private static final String ROOT = "confetti";
+
+    public static void registerCorePermissions() {
+        Permission parent = DefaultPermissions.registerPermission(ROOT, "Gives the user the ability to use all Confetti utilities and commands");
+
+        DefaultPermissions.registerPermission(ROOT + ".maximumtrackerbypass", "Gives the user the ability to see all entities regardless of the maximum tracker config", PermissionDefault.OP, parent);
+
+        ConfettiCommandPermissions.registerPermissions(parent);
+
+        parent.recalculatePermissibles();
+    }
+}
diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index aad2cdf4c2237caf2c99695d0cbb01f56c0074e2..bbe72c8b2ebe5ba2458b4f75a1f0fd2a8ed5d4c9 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -23,7 +23,6 @@ import joptsimple.util.PathConverter;
 import joptsimple.util.PathProperties;
 import net.minecraft.CrashReport;
 import net.minecraft.DefaultUncaughtExceptionHandler;
-import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.commands.Commands;
 import net.minecraft.core.Registry;
@@ -69,6 +68,9 @@ import net.minecraft.SharedConstants;
 import org.bukkit.configuration.file.YamlConfiguration;
 // CraftBukkit end
 
+import com.vulpeslab.confetti.config.ConfettiConfiguration;
+import com.vulpeslab.confetti.config.ConfettiConfigurationLoader;
+
 public class Main {
 
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -120,6 +122,8 @@ public class Main {
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
+            ConfettiConfigurationLoader.init((File) optionset.valueOf("confetti-settings")); // Confetti
+
             // Purpur start - load config files early
             org.bukkit.configuration.file.YamlConfiguration purpurConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) optionset.valueOf("purpur-settings"));
             org.purpurmc.purpur.PurpurConfig.clampEnchantLevels = purpurConfiguration.getBoolean("settings.enchantment.clamp-levels");
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 8854e3adca98214697fdfa7f1296e2d66b5afa48..0218475c1607d620c778ad2f0c1997e536d7f751 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -261,6 +261,7 @@ import org.yaml.snakeyaml.constructor.SafeConstructor;
 import org.yaml.snakeyaml.error.MarkedYAMLException;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import com.vulpeslab.confetti.permissions.ConfettiPermissions;
 
 import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
@@ -614,6 +615,7 @@ public final class CraftServer implements Server {
             this.commandMap.registerServerAliases();
             DefaultPermissions.registerCorePermissions();
             CraftDefaultPermissions.registerCorePermissions();
+            ConfettiPermissions.registerCorePermissions(); // Confetti
             if (!io.papermc.paper.configuration.GlobalConfiguration.get().misc.loadPermissionsYmlBeforePlugins) this.loadCustomPermissions(); // Paper
             this.helpMap.initializeCommands();
             this.syncCommands();
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 068aa459b4d1dfbbbd80dc0ae5c43b8202474477..e9dfab2d5871eaca59cb10faf0ae8e10bb46feec 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -190,6 +190,14 @@ public class Main {
                         .describedAs("Yml file");
                 // Purpur end
 
+                // Confetti start
+                acceptsAll(asList("confetti", "confetti-settings"), "File for confetti settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("confetti.yml"))
+                        .describedAs("Yml file");
+                // Confetti end
+
                 // Paper start
                 acceptsAll(asList("server-name"), "Name of the server")
                         .withRequiredArg()
