From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 4 May 2024 14:40:58 +0900
Subject: [PATCH] Add multithreading region scheduler


diff --git a/src/main/java/com/vulpeslab/confetti/region/RegionPos.java b/src/main/java/com/vulpeslab/confetti/region/RegionPos.java
new file mode 100644
index 0000000000000000000000000000000000000000..64231d4c3fcee32f8243cc0d7c75795c89b55a78
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/region/RegionPos.java
@@ -0,0 +1,138 @@
+package com.vulpeslab.confetti.region;
+
+import com.mojang.logging.LogUtils;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.SectionPos;
+import net.minecraft.world.level.ChunkPos;
+import org.bukkit.Location;
+import org.slf4j.Logger;
+import com.vulpeslab.confetti.config.ConfettiConfiguration;
+
+// Copy of ChunkPos basically, but separate to avoid accidental usage as a ChunkPos
+public class RegionPos {
+
+    private static final Logger LOGGER = LogUtils.getClassLogger();
+
+    public static final int REGION_SIZE; // eg 8 (for an 8x8 region)
+    public static final int REGION_SHIFT; // eg 3 (1 << 3 == 8)
+    public static final int REGION_SIZE_MASK; // eg 7 (9 % 8 == 9 & 7 == 1)
+
+    static {
+        // desiredRegionSize = 7 -> shift = 3, size = 8, mask = 7
+        // desiredRegionSize = 8 -> shift = 3, size = 8, mask = 7
+        // desiredRegionSize = 9 -> shift = 4, size = 16, mask = 15
+        int desiredRegionSize = ConfettiConfiguration.get().multithreading.regionSize - 1;
+        int shift = 0;
+        while (desiredRegionSize > 0) {
+            shift++;
+            desiredRegionSize >>= 1;
+        }
+
+        REGION_SIZE = 1 << shift;
+        REGION_SHIFT = shift;
+        REGION_SIZE_MASK = (1 << shift) - 1;
+
+        if (REGION_SIZE <= 1) {
+            throw new IllegalStateException("!!! Region size is " + REGION_SIZE + " chunk, this is too small. It must be at least 2 chunks !!!");
+        }
+
+        if (REGION_SIZE < 8) {
+            LOGGER.warn("!!! !!!");
+            LOGGER.warn("!!! Region size is less than 8 chunks, this will cause issues unless you know what you're doing!!!");
+            LOGGER.warn("!!! !!!");
+        }
+
+        LOGGER.info("Using region size: {}, shift={}, mask={}", REGION_SIZE, REGION_SHIFT, REGION_SIZE_MASK);
+    }
+
+    public final int x;
+    public final int z;
+    public final long longKey; // Paper
+
+    public RegionPos(int rx, int rz) {
+        this.x = rx;
+        this.z = rz;
+        this.longKey = RegionPos.asLong(this.x, this.z);
+    }
+
+    public RegionPos(long regionKey) {
+        this.x = (int) regionKey;
+        this.z = (int) (regionKey >> 32);
+        this.longKey = regionKey;
+    }
+
+    public static RegionPos forChunk(ChunkPos chunkPos) {
+        return chunkPos.getRegionPos(); // Cache the RegionPos on the ChunkPos to minimize object creation
+    }
+
+    public static RegionPos forChunk(int chunkX, int chunkZ) {
+        return new RegionPos(asLongForChunk(chunkX, chunkZ));
+    }
+
+    public static RegionPos forBlockPos(BlockPos blockPos) {
+        return forBlockPos(blockPos.getX(), blockPos.getY(), blockPos.getZ());
+    }
+
+    public static RegionPos forBlockPos(int x, int y, int z) {
+        return new RegionPos(asLongForBlockPos(x, y, z));
+    }
+
+    public static RegionPos forLocation(Location location) {
+        return forBlockPos(location.getBlockX(), location.getBlockY(), location.getBlockZ());
+    }
+
+    public static long asLongForChunk(ChunkPos chunkPos) {
+        return asLongForChunk(chunkPos.x, chunkPos.z);
+    }
+
+    public static long asLongForChunk(int chunkX, int chunkZ) {
+        return asLong(chunkX >> REGION_SHIFT, chunkZ >> REGION_SHIFT);
+    }
+
+    public static long asLongForBlockPos(BlockPos blockPos) {
+        return asLongForBlockPos(blockPos.getX(), blockPos.getY(), blockPos.getZ());
+    }
+
+    public static long asLongForBlockPos(int x, int y, int z) {
+        return asLongForChunk(SectionPos.blockToSectionCoord(x), SectionPos.blockToSectionCoord(z));
+    }
+
+    public static long asLong(RegionPos regionPos) {
+        return asLong(regionPos.x, regionPos.z);
+    }
+
+    public static long asLong(int regionX, int regionZ) {
+        return (long)regionX & 4294967295L | ((long)regionZ & 4294967295L) << 32;
+    }
+
+    public int getLowerChunkX() {
+        return this.x << REGION_SHIFT;
+    }
+
+    public int getLowerChunkZ() {
+        return this.z << REGION_SHIFT;
+    }
+
+    public int getUpperChunkX() {
+        return getLowerChunkX() + REGION_SIZE - 1;
+    }
+
+    public int getUpperChunkZ() {
+        return getLowerChunkZ() + REGION_SIZE - 1;
+    }
+
+    @Override
+    public int hashCode() {
+        return ChunkPos.hash(this.x, this.z);
+    }
+
+    @Override
+    public boolean equals(Object object) {
+        return this == object || object instanceof RegionPos regionPos && this.x == regionPos.x && this.z == regionPos.z;
+    }
+
+    @Override
+    public String toString() {
+        return "RegionPos[" + this.x + ", " + this.z + "]";
+    }
+}
diff --git a/src/main/java/com/vulpeslab/confetti/threading/ConfettiRegionLocker.java b/src/main/java/com/vulpeslab/confetti/threading/ConfettiRegionLocker.java
new file mode 100644
index 0000000000000000000000000000000000000000..5a7ecda3ac5ab2c1ea36eb7dc90ec9076008dce1
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/threading/ConfettiRegionLocker.java
@@ -0,0 +1,320 @@
+package com.vulpeslab.confetti.threading;
+
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import com.vulpeslab.confetti.region.RegionPos;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.LockSupport;
+import java.util.function.Supplier;
+
+/**
+ * Read lock:
+ *  - Only one thread can hold a read lock at a time.
+ *  - Many servers can hold a read lock at the same time.
+ *  - Good for tasks such as sending data to clients and loading chunks.
+ * Write lock:
+ *  - An extension of the read lock.
+ *  - Only one server can hold a write lock at a time.
+ *  - Important for tasks that modify the region, such as modifying blocks or entities.
+ */
+public class ConfettiRegionLocker {
+
+    private static final int REGION_LOCK_RADIUS = 1;
+
+    private final ConcurrentHashMap<RegionPos, LockedRegion> lockedRegions = new ConcurrentHashMap<>();
+
+    private final ThreadLocal<Set<RegionPos>> localLocks = ThreadLocal.withInitial(ObjectOpenHashSet::new);
+    private final ThreadLocal<Set<RegionPos>> readOnlyLocks = ThreadLocal.withInitial(ObjectOpenHashSet::new);
+    private final ThreadLocal<Set<RegionPos>> writeLocks = ThreadLocal.withInitial(ObjectOpenHashSet::new);
+    private final ThreadLocal<Set<RegionPos>> unmodifiableLocalLocks = ThreadLocal.withInitial(() -> Collections.unmodifiableSet(localLocks.get()));
+
+    /**
+     * Checks if the current thread holds a read lock for the given region
+     */
+    public boolean hasLock(RegionPos regionPos) {
+        return localLocks.get().contains(regionPos);
+    }
+
+    /**
+     * Checks if the current thread holds a write lock for the given region.
+     * This check is usually unnecessary, but ensures that there will be no
+     * syncing conflicts with other servers.
+     */
+    public boolean hasWriteLock(RegionPos regionPos) {
+        return writeLocks.get().contains(regionPos);
+    }
+
+    /**
+     * Returns an unmodifiable view of the locked regions for the current thread.
+     */
+    public Set<RegionPos> getLockedRegions() {
+        // Use an unmodifiable view to ensure the underlying set doesn't get accidentally modified
+        return unmodifiableLocalLocks.get();
+    }
+
+    /**
+     * Returns an unmodifiable view of all locked regions across all threads.
+     */
+    public Set<Map.Entry<RegionPos, LockedRegion>> getAllLockedRegions() {
+        return Collections.unmodifiableSet(lockedRegions.entrySet());
+    }
+
+    public RegionLock lockRegion(RegionPos regionPos) {
+        int tries = 0;
+        RegionLock lock;
+        while ((lock = this.tryTakeLockNow(regionPos)) == null) {
+            LockSupport.parkNanos(Math.min(100_000, 1_000 * (++tries)));
+        }
+        return lock;
+    }
+
+    /**
+     * Lock the region and run the runnable. If the region is already locked, wait until it is unlocked.
+     * Can be called recursively and will respect existing locks created by the same thread.
+     */
+    public <T> T lockRegion(RegionPos regionPos, Supplier<T> runnableWithReturnValue) {
+        RegionLock lock = this.lockRegion(regionPos);
+        try {
+            return runnableWithReturnValue.get();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Lock the region and run the runnable. If the region is already locked, wait until it is unlocked.
+     * Can be called recursively and will respect existing locks created by the same thread.
+     */
+    public void lockRegion(RegionPos regionPos, Runnable runnable) {
+        RegionLock lock = this.lockRegion(regionPos);
+        try {
+            runnable.run();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Try to acquire the region lock immediately, if successful run the runnable.
+     * If unsuccessful, return false and the runnable will not be run.
+     * If the region is already locked, it will return unsuccessfully immediately instead of waiting to try to acquire the lock.
+     * Can be called recursively and will respect existing locks created by the same thread.
+     * This method will sync the lock with other servers.
+     * @return true if the lock was acquired and the runnable was run, false if the runnable was not run
+     */
+    public boolean tryLockNow(RegionPos centerPos, Runnable ifSuccess) {
+        RegionLock lock = this.tryTakeLockNow(centerPos);
+        if (lock == null) {
+            return false;
+        }
+
+        try {
+            ifSuccess.run();
+            return true;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Try to acquire the region lock immediately, if successful run the runnable.
+     * If unsuccessful, return false and the runnable will not be run.
+     * If the region is already locked, it will return unsuccessfully immediately instead of waiting to try to acquire the lock.
+     * Can be called recursively and will respect existing locks created by the same thread.
+     * The specified region must not be modified within this lock. This method will not sync the lock with other servers.
+     * return true if the lock was acquired and the runnable was run, false if the runnable was not run
+     */
+    public boolean tryReadOnlyLockNow(RegionPos centerPos, Runnable ifSuccess) {
+        final RegionLock lock = this.tryTakeReadOnlyLockNow(centerPos);
+        if (lock == null) {
+            return false;
+        }
+
+        try {
+            ifSuccess.run();
+            return true;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Nullable
+    public RegionLock tryTakeLockNow(RegionPos centerPos) {
+        return internalTryTakeLockNow(centerPos, REGION_LOCK_RADIUS);
+    }
+
+    @Nullable
+    public WriteRegionLock internalTryTakeLockNow(RegionPos centerPos, int lockRadius) {
+        final ReadOnlyRegionLock lock = internalTryTakeReadOnlyLockNow(centerPos, lockRadius);
+        return lock == null ? null : new WriteRegionLock(lock);
+    }
+
+    @Nullable
+    public RegionLock tryTakeReadOnlyLockNow(RegionPos centerPos) {
+        return internalTryTakeReadOnlyLockNow(centerPos, REGION_LOCK_RADIUS);
+    }
+
+    @Nullable
+    public ReadOnlyRegionLock internalTryTakeReadOnlyLockNow(RegionPos centerPos, int lockRadius) {
+        final ReadOnlyRegionLock lock = new ReadOnlyRegionLock(lockRadius);
+
+        for (int x = -lockRadius; x <= lockRadius; x++) {
+            for (int z = -lockRadius; z <= lockRadius; z++) {
+                RegionPos regionPos = x == 0 && z == 0 ? centerPos : new RegionPos(centerPos.x + x, centerPos.z + z);
+                if (!lock.tryLockRegion(regionPos)) {
+                    // Failed to take lock, abort
+                    lock.unlock();
+                    return null;
+                }
+            }
+        }
+
+        return lock;
+    }
+
+    @Nullable
+    public CompletableFuture<Void> onUnlock(RegionPos centerPos, Supplier<CompletableFuture<Void>> nextTask) {
+        for (int x = -REGION_LOCK_RADIUS; x <= REGION_LOCK_RADIUS; x++) {
+            for (int z = -REGION_LOCK_RADIUS; z <= REGION_LOCK_RADIUS; z++) {
+                RegionPos regionPos = x == 0 && z == 0 ? centerPos : new RegionPos(centerPos.x + x, centerPos.z + z);
+                LockedRegion lockedRegion = this.lockedRegions.get(regionPos);
+                if (lockedRegion != null) {
+                    return lockedRegion.onUnlock(nextTask);
+                }
+            }
+        }
+
+        return nextTask.get();
+    }
+
+    public interface RegionLock {
+        Collection<RegionPos> lockedRegions();
+        Thread owner();
+        void unlock();
+    }
+
+    public class ReadOnlyRegionLock implements RegionLock {
+        private final List<LockedRegion> readLocks;
+        private final Thread thread;
+
+        private ReadOnlyRegionLock(int lockRadius) {
+            this.thread = Thread.currentThread();
+            this.readLocks = new ArrayList<>((lockRadius * 2 + 1) * (lockRadius * 2 + 1));
+        }
+
+        protected boolean tryLockRegion(RegionPos regionPos) {
+            if (this.thread != Thread.currentThread()) {
+                throw new IllegalStateException("Cannot lock a region from a different thread [expected=%s,got=%s]".formatted(thread, Thread.currentThread()));
+            }
+
+            LockedRegion lockedRegion = ConfettiRegionLocker.this.lockedRegions.compute(regionPos, (k, prevValue) -> {
+                if (prevValue == null) {
+                    // This region is unlocked, let's lock it
+                    ConfettiRegionLocker.this.localLocks.get().add(regionPos);
+                    ConfettiRegionLocker.this.readOnlyLocks.get().add(regionPos);
+                    LockedRegion newLockedRegion = new LockedRegion(regionPos, this.thread);
+                    this.readLocks.add(newLockedRegion);
+                    return newLockedRegion;
+                } else {
+                    // This region is already locked, it could be already locked by us or someone else
+                    return prevValue;
+                }
+            });
+
+            return lockedRegion.owner() == this.owner();
+        }
+
+        public Collection<RegionPos> lockedRegions() {
+            return this.readLocks.stream().map(LockedRegion::regionPos).toList();
+        }
+
+        public Thread owner() {
+            return this.thread;
+        }
+
+        public void unlock() {
+            if (this.thread != Thread.currentThread()) {
+                throw new IllegalStateException("Cannot unlock a lock from a different thread [expected=%s,got=%s]".formatted(thread, Thread.currentThread()));
+            }
+
+            for (LockedRegion lockedRegion : this.readLocks) {
+                ConfettiRegionLocker.this.lockedRegions.remove(lockedRegion.regionPos(), lockedRegion);
+                ConfettiRegionLocker.this.localLocks.get().remove(lockedRegion.regionPos());
+                ConfettiRegionLocker.this.readOnlyLocks.get().remove(lockedRegion.regionPos());
+            }
+
+            for (LockedRegion lockedRegion : this.readLocks) {
+                lockedRegion.complete();
+            }
+        }
+
+    }
+
+    public class WriteRegionLock implements RegionLock {
+        private final List<RegionPos> writeLocks;
+        private final ReadOnlyRegionLock superLock;
+
+        private WriteRegionLock(ReadOnlyRegionLock superLock) {
+            this.superLock = superLock;
+
+            Collection<RegionPos> lockedRegions = superLock.lockedRegions();
+            this.writeLocks = new ArrayList<>(lockedRegions);
+            ConfettiRegionLocker.this.writeLocks.get().addAll(lockedRegions);
+            ConfettiRegionLocker.this.readOnlyLocks.get().removeAll(lockedRegions);
+        }
+
+        @Override
+        public Collection<RegionPos> lockedRegions() {
+            return Collections.unmodifiableCollection(this.writeLocks);
+        }
+
+        @Override
+        public Thread owner() {
+            return this.superLock.owner();
+        }
+
+        @Override
+        public void unlock() {
+            ConfettiRegionLocker.this.writeLocks.get().removeAll(this.writeLocks);
+            ConfettiRegionLocker.this.readOnlyLocks.get().addAll(this.writeLocks);
+            this.superLock.unlock();
+        }
+    }
+
+    public static class LockedRegion {
+        private final RegionPos regionPos;
+        private final Thread owner;
+        private final CompletableFuture<Void> headUnlockFuture = new CompletableFuture<>();
+        private CompletableFuture<Void> tailUnlockFuture = headUnlockFuture;
+
+        private LockedRegion(RegionPos regionPos, Thread owner) {
+            this.regionPos = regionPos;
+            this.owner = owner;
+        }
+
+        public RegionPos regionPos() {
+            return regionPos;
+        }
+
+        public Thread owner() {
+            return owner;
+        }
+
+        public void complete() {
+            headUnlockFuture.complete(null);
+        }
+
+        public CompletableFuture<Void> onUnlock(Supplier<CompletableFuture<Void>> nextTask) {
+            return tailUnlockFuture = tailUnlockFuture.thenCompose(v -> nextTask.get());
+        }
+    }
+
+}
diff --git a/src/main/java/com/vulpeslab/confetti/threading/ConfettiRegionScheduler.java b/src/main/java/com/vulpeslab/confetti/threading/ConfettiRegionScheduler.java
new file mode 100644
index 0000000000000000000000000000000000000000..bcaa52a2316d046f380ecfbd78c229f9213b0faa
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/threading/ConfettiRegionScheduler.java
@@ -0,0 +1,185 @@
+package com.vulpeslab.confetti.threading;
+
+import net.minecraft.server.level.ServerLevel;
+import com.vulpeslab.confetti.region.RegionPos;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+public class ConfettiRegionScheduler {
+
+    private final ConfettiRegionLocker locker = new ConfettiRegionLocker();
+
+    public static CompletableFuture<Void> schedule(ServerLevel level, RegionPos regionPos, Runnable runnable) {
+        return level.chunkScheduler.schedule(regionPos, runnable);
+    }
+
+    /**
+     * Schedule a task to run on the given region's thread.
+     */
+    public CompletableFuture<Void> schedule(RegionPos regionPos, Runnable runnable) {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+
+        ConfettiTickThread.getExecutor().execute(() -> {
+            run(regionPos, runnable, future);
+        });
+
+        return future;
+    }
+
+    /**
+     * Avoid using this often. Locking massive parts of the world can take time and will freeze these regions during the process.
+     */
+    public CompletableFuture<Void> scheduleOnMany(Runnable runnable, RegionPos... posArray) {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+
+        ConfettiTickThread.getExecutor().execute(() -> {
+            runOnMany(sortPredictably(posArray), runnable, future);
+        });
+
+        return future;
+    }
+
+    public static CompletableFuture<Void> scheduleAcrossLevels(ServerLevel level1, RegionPos regionPos1, ServerLevel level2, RegionPos regionPos2, Runnable runnable) {
+        if (level1 == level2) {
+            // We don't sort the regionPos in this method because we assume they're a different level, use the method that does sort them instead
+            return level1.chunkScheduler.scheduleOnMany(runnable, regionPos1, regionPos2);
+        }
+
+        ServerLevel finalLevel1;
+        RegionPos finalRegionPos1;
+        ServerLevel finalLevel2;
+        RegionPos finalRegionPos2;
+
+        // Sort predictably to avoid deadlocks
+        if (compare(level1, level2) > 0) {
+            finalLevel1 = level2;
+            finalRegionPos1 = regionPos2;
+            finalLevel2 = level1;
+            finalRegionPos2 = regionPos1;
+        } else {
+            finalLevel1 = level1;
+            finalRegionPos1 = regionPos1;
+            finalLevel2 = level2;
+            finalRegionPos2 = regionPos2;
+        }
+
+        CompletableFuture<Void> future = new CompletableFuture<>();
+
+        ConfettiTickThread.getExecutor().execute(() -> {
+            runAcrossLevels(finalLevel1, finalRegionPos1, finalLevel2, finalRegionPos2, runnable, future);
+        });
+
+        return future;
+    }
+
+    /**
+     * Sort the region positions in a predictable order to avoid deadlocks.
+     */
+    private RegionPos[] sortPredictably(RegionPos[] posArray) {
+        // This should only be used on very small arrays, usually just 2 elements, so this simple O(n^2) sort is fine
+        for (int i = 0; i < posArray.length; i++) {
+            for (int j = i + 1; j < posArray.length; j++) {
+                if (compare(posArray[i], posArray[j]) > 0) {
+                    RegionPos temp = posArray[i];
+                    posArray[i] = posArray[j];
+                    posArray[j] = temp;
+                }
+            }
+        }
+        return posArray;
+    }
+
+    private static int compare(RegionPos a, RegionPos b) {
+        int x = Integer.compare(a.x, b.x);
+        if (x != 0) {
+            return x;
+        }
+        return Integer.compare(a.z, b.z);
+    }
+
+    private static int compare(ServerLevel a, ServerLevel b) {
+        return a.uuid.compareTo(b.uuid);
+    }
+
+    private void run(RegionPos regionPos, Runnable runnable, CompletableFuture<Void> future) {
+        ConfettiRegionLocker.RegionLock lock = null;
+        try {
+            lock = locker.tryTakeLockNow(regionPos);
+            if (lock == null) {
+                // Wait for unlock, then retry
+                locker.onUnlock(regionPos, () -> CompletableFuture.runAsync(() -> run(regionPos, runnable, future), ConfettiTickThread.getExecutor()));
+                return;
+            }
+
+            try {
+                runnable.run();
+            } finally {
+                lock.unlock();
+            }
+
+            future.complete(null);
+        } catch (Exception e) {
+            future.completeExceptionally(e);
+        }
+    }
+
+    private void runOnMany(RegionPos[] regionPosArray, Runnable runnable, CompletableFuture<Void> future) {
+        ConfettiRegionLocker.RegionLock[] locks = new ConfettiRegionLocker.RegionLock[regionPosArray.length];
+        try {
+            try {
+                for (int i = 0; i < regionPosArray.length; i++) {
+                    locks[i] = locker.tryTakeLockNow(regionPosArray[i]);
+                    if (locks[i] == null) {
+                        locker.onUnlock(regionPosArray[i], () -> CompletableFuture.runAsync(() -> runOnMany(regionPosArray, runnable, future), ConfettiTickThread.getExecutor()));
+                        return;
+                    }
+                }
+
+                // All locks acquired, run the task
+                runnable.run();
+            } finally {
+                for (int i = locks.length - 1; i >= 0; i--) {
+                    if (locks[i] != null) locks[i].unlock();
+                }
+            }
+
+            future.complete(null);
+        } catch (Exception e) {
+            future.completeExceptionally(e);
+        }
+    }
+
+    private static void runAcrossLevels(ServerLevel level1, RegionPos regionPos1, ServerLevel level2, RegionPos regionPos2, Runnable runnable, CompletableFuture<Void> future) {
+        ConfettiRegionLocker.RegionLock lock1 = null;
+        ConfettiRegionLocker.RegionLock lock2 = null;
+        try {
+            try {
+                lock1 = level1.chunkScheduler.locker.tryTakeLockNow(regionPos1);
+                lock2 = lock1 == null ? null : level2.chunkScheduler.locker.tryTakeLockNow(regionPos2);
+
+                if (lock2 == null) {
+                    Supplier<CompletableFuture<Void>> unlockCallback = () -> CompletableFuture.runAsync(() -> runAcrossLevels(level1, regionPos1, level2, regionPos2, runnable, future), ConfettiTickThread.getExecutor());
+                    if (lock1 == null) level1.chunkScheduler.locker.onUnlock(regionPos1, unlockCallback);
+                    else level2.chunkScheduler.locker.onUnlock(regionPos2, unlockCallback);
+                    return;
+                }
+
+                // Both locks acquired, run the task
+                runnable.run();
+            } finally {
+                if (lock2 != null) lock2.unlock();
+                if (lock1 != null) lock1.unlock();
+            }
+
+            future.complete(null);
+        } catch (Exception e) {
+            future.completeExceptionally(e);
+        }
+    }
+
+    public ConfettiRegionLocker getRegionLocker() {
+        return locker;
+    }
+
+}
diff --git a/src/main/java/com/vulpeslab/confetti/threading/ConfettiTickThread.java b/src/main/java/com/vulpeslab/confetti/threading/ConfettiTickThread.java
new file mode 100644
index 0000000000000000000000000000000000000000..51719858bdf39447e37b9950473f3a45a3a90aa9
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/threading/ConfettiTickThread.java
@@ -0,0 +1,62 @@
+package com.vulpeslab.confetti.threading;
+
+import com.mojang.logging.LogUtils;
+import io.papermc.paper.util.TickThread;
+import org.slf4j.Logger;
+import com.vulpeslab.confetti.config.ConfettiConfiguration;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ConfettiTickThread extends TickThread {
+
+    private static final Logger LOGGER = LogUtils.getClassLogger();
+
+    public static final int THREAD_COUNT;
+
+    private static final AtomicInteger ID_GENERATOR = new AtomicInteger();
+
+    private static final ExecutorService executor;
+
+    static {
+        int threadCount = ConfettiConfiguration.get().multithreading.threadCount;
+
+        if (threadCount <= 0) {
+            threadCount = Math.max(Runtime.getRuntime().availableProcessors() - 1, 1);
+        }
+
+        THREAD_COUNT = threadCount;
+
+        executor = Executors.newFixedThreadPool(THREAD_COUNT, r -> new ConfettiTickThread(r, "ConfettiTickThread-%d"));
+
+        LOGGER.info("Using {} threads", THREAD_COUNT);
+    }
+
+    public ConfettiTickThread(Runnable run, String name) {
+        super(run, String.format(name, ID_GENERATOR.incrementAndGet()));
+    }
+
+    public static boolean isConfettiTickThread() {
+        // Use this method to check if it's a confetti tick thread, to ensure future potential support for VirtualThreads
+        return Thread.currentThread() instanceof ConfettiTickThread;
+    }
+
+    public static ExecutorService getExecutor() {
+        return executor;
+    }
+
+    public static void stopServer() {
+        executor.shutdown();
+        try {
+            executor.awaitTermination(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+        if (!executor.isTerminated()) {
+            LOGGER.warn("Failed to stop tick threads after 5 seconds. Terminating...");
+            executor.shutdownNow();
+        }
+    }
+}
diff --git a/src/main/java/com/vulpeslab/confetti/util/ObjectHolder.java b/src/main/java/com/vulpeslab/confetti/util/ObjectHolder.java
new file mode 100644
index 0000000000000000000000000000000000000000..086dc310ee02f172e1bb76257986b5e4901e92e8
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/util/ObjectHolder.java
@@ -0,0 +1,22 @@
+package com.vulpeslab.confetti.util;
+
+public class ObjectHolder <T> {
+
+    T value;
+
+    public ObjectHolder() {
+        this(null);
+    }
+
+    public ObjectHolder(T object) {
+        this.value = object;
+    }
+
+    public T value() {
+        return value;
+    }
+
+    public void value(T value) {
+        this.value = value;
+    }
+}
diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
index 7ce9ebba8ce304d1f3f21d4f15ee5f3560d7700b..42df50c96e78100ab49afcf35b4c8d142223ad41 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
@@ -4,6 +4,8 @@ import co.aikar.timings.TimedEventExecutor;
 import com.destroystokyo.paper.event.server.ServerExceptionEvent;
 import com.destroystokyo.paper.exception.ServerEventException;
 import com.google.common.collect.Sets;
+import com.vulpeslab.confetti.threading.ConfettiTickThread;
+import io.papermc.paper.util.TickThread;
 import org.bukkit.Server;
 import org.bukkit.Warning;
 import org.bukkit.event.Event;
@@ -36,7 +38,7 @@ class PaperEventManager {
 
     // SimplePluginManager
     public void callEvent(@NotNull Event event) {
-        if (event.isAsynchronous() && this.server.isPrimaryThread()) {
+        if (event.isAsynchronous() && (Thread.currentThread() instanceof TickThread || ConfettiTickThread.isConfettiTickThread())) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
         } else if (!event.isAsynchronous() && !this.server.isPrimaryThread() && !this.server.isStopping()) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 1d35eb164d03dbe16b244fb2f26c6b13a61b9360..33808b1b21c656672c5e5a57ef40668cf44ebf04 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -197,6 +197,7 @@ import org.bukkit.event.server.ServerLoadEvent;
 // CraftBukkit end
 
 import co.aikar.timings.MinecraftTimings; // Paper
+import com.vulpeslab.confetti.threading.ConfettiTickThread; // Confetti
 
 public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTask> implements ServerInfo, ChunkIOErrorReporter, CommandSource, AutoCloseable, ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer { // Paper - rewrite chunk system
 
@@ -1045,6 +1046,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         MinecraftServer.LOGGER.info("Stopping server");
         Commands.COMMAND_SENDING_POOL.shutdownNow(); // Paper - Perf: Async command map building; Shutdown and don't bother finishing
         MinecraftTimings.stopServer(); // Paper
+        ConfettiTickThread.stopServer(); // Confetti
         this.server.spark.disable(); // Paper - spark
         // Purpur start
         if (upnp) {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java.rej b/src/main/java/net/minecraft/server/MinecraftServer.java.rej
index ff3c9c601ab6c5e8053621137084f27938cd2922..f2b5e6ea78373827239b8dd7472a1f45fc7820ce 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java.rej
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java.rej
@@ -1,10 +1,17 @@
 diff a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java	(rejected hunks)
-@@ -960,7 +960,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
-         shutdownThread = Thread.currentThread();
-         org.spigotmc.WatchdogThread.doStop(); // Paper
-         if (!isSameThread()) {
--            MinecraftServer.LOGGER.info("Stopping main thread (Ignore any thread death message you see! - DO NOT REPORT THREAD DEATH TO PURPUR)"); // Purpur
-+            MinecraftServer.LOGGER.info("Stopping main thread (Ignore any thread death message you see! - DO NOT REPORT THREAD DEATH TO CONFETTI)"); // Purpur // Confetti
-             while (this.getRunningThread().isAlive()) {
-                 this.getRunningThread().stop();
-                 try {
+@@ -191,6 +191,7 @@ import org.bukkit.event.server.ServerLoadEvent;
+ // CraftBukkit end
+ 
+ import co.aikar.timings.MinecraftTimings; // Paper
++import com.vulpeslab.confetti.threading.ConfettiTickThread;
+ 
+ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTask> implements ServerInfo, CommandSource, AutoCloseable {
+ 
+@@ -977,6 +978,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
+         MinecraftServer.LOGGER.info("Stopping server");
+         Commands.COMMAND_SENDING_POOL.shutdownNow(); // Paper - Perf: Async command map building; Shutdown and don't bother finishing
+         MinecraftTimings.stopServer(); // Paper
++        ConfettiTickThread.stopServer(); // Confetti
+         // Purpur start
+         if (upnp) {
+             if (dev.omega24.upnp4j.UPnP4J.close(this.getPort(), dev.omega24.upnp4j.util.Protocol.TCP)) {
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index e6715ac13d573a01eb0307ec2417a430547222a6..a616991e06049c0c25ec2ee8cfd3e8181d73aaba 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -51,7 +51,6 @@ import net.minecraft.core.particles.ParticleOptions;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.core.registries.Registries;
 import net.minecraft.network.chat.Component;
-import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ClientboundBlockDestructionPacket;
 import net.minecraft.network.protocol.game.ClientboundBlockEventPacket;
@@ -79,8 +78,6 @@ import net.minecraft.util.Mth;
 import net.minecraft.util.ProgressListener;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.Unit;
-import net.minecraft.util.datafix.DataFixTypes;
-import net.minecraft.util.profiling.ProfilerFiller;
 import net.minecraft.util.valueproviders.IntProvider;
 import net.minecraft.util.valueproviders.UniformInt;
 import net.minecraft.world.DifficultyInstance;
@@ -135,16 +132,13 @@ import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.chunk.LevelChunkSection;
 import net.minecraft.world.level.chunk.storage.EntityStorage;
 import net.minecraft.world.level.chunk.storage.RegionStorageInfo;
-import net.minecraft.world.level.chunk.storage.SimpleRegionStorage;
 import net.minecraft.world.level.dimension.BuiltinDimensionTypes;
 import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.dimension.end.EndDragonFight;
-import net.minecraft.world.level.entity.EntityPersistentStorage;
 import net.minecraft.world.level.entity.EntityTickList;
 import net.minecraft.world.level.entity.EntityTypeTest;
 import net.minecraft.world.level.entity.LevelCallback;
 import net.minecraft.world.level.entity.LevelEntityGetter;
-import net.minecraft.world.level.entity.PersistentEntitySectionManager;
 import net.minecraft.world.level.gameevent.DynamicGameEventListener;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.gameevent.GameEventDispatcher;
@@ -181,6 +175,7 @@ import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.bukkit.event.server.MapInitializeEvent;
 import org.bukkit.event.weather.LightningStrikeEvent;
 import org.bukkit.event.world.TimeSkipEvent;
+import com.vulpeslab.confetti.threading.ConfettiRegionScheduler; // Confetti
 // CraftBukkit end
 
 public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel, ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevelReader { // Paper - rewrite chunk system
@@ -224,6 +219,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
     private double preciseTime; // Purpur
     private boolean forceTime; // Purpur
     private final RandomSequences randomSequences;
+    public final ConfettiRegionScheduler chunkScheduler = new ConfettiRegionScheduler(); // Confetti
 
     // CraftBukkit start
     public final LevelStorageSource.LevelStorageAccess convertable;
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java.rej b/src/main/java/net/minecraft/server/level/ServerLevel.java.rej
new file mode 100644
index 0000000000000000000000000000000000000000..53ffefd925878b67cd78486faed07ecc257e1101
--- /dev/null
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java.rej
@@ -0,0 +1,17 @@
+diff a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java	(rejected hunks)
+@@ -180,6 +174,7 @@ import org.bukkit.event.entity.CreatureSpawnEvent;
+ import org.bukkit.event.server.MapInitializeEvent;
+ import org.bukkit.event.weather.LightningStrikeEvent;
+ import org.bukkit.event.world.TimeSkipEvent;
++import com.vulpeslab.confetti.threading.ConfettiRegionScheduler;
+ // CraftBukkit end
+ 
+ public class ServerLevel extends Level implements WorldGenLevel {
+@@ -223,6 +218,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
+     private double preciseTime; // Purpur
+     private boolean forceTime; // Purpur
+     private final RandomSequences randomSequences;
++    public final ConfettiRegionScheduler chunkScheduler = new ConfettiRegionScheduler(); // Confetti
+     public long lastMidTickExecuteFailure; // Paper - execute chunk tasks mid tick
+ 
+     // CraftBukkit start
diff --git a/src/main/java/net/minecraft/world/level/ChunkPos.java b/src/main/java/net/minecraft/world/level/ChunkPos.java
index fa58eeec2b652f0fa251eedf11cfabde5fd3198b..8aa27795db261da86447879e0b4a0d70e6b09ffe 100644
--- a/src/main/java/net/minecraft/world/level/ChunkPos.java
+++ b/src/main/java/net/minecraft/world/level/ChunkPos.java
@@ -7,6 +7,7 @@ import java.util.stream.StreamSupport;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
+import com.vulpeslab.confetti.region.RegionPos;
 
 public class ChunkPos {
     private static final int SAFETY_MARGIN = 1056;
@@ -24,6 +25,7 @@ public class ChunkPos {
     private static final int HASH_A = 1664525;
     private static final int HASH_C = 1013904223;
     private static final int HASH_Z_XOR = -559038737;
+    @Nullable private RegionPos regionPos; // Confetti
 
     public ChunkPos(int x, int z) {
         this.x = x;
@@ -43,6 +45,16 @@ public class ChunkPos {
         this.longKey = asLong(this.x, this.z); // Paper
     }
 
+    // Confetti start
+    public RegionPos getRegionPos() {
+        if (this.regionPos == null) {
+            // Cache the RegionPos to minimize object creation
+            this.regionPos = RegionPos.forChunk(this.x, this.z);
+        }
+        return this.regionPos;
+    }
+    // Confetti end
+
     public static ChunkPos minFromRegion(int x, int z) {
         return new ChunkPos(x << 5, z << 5);
     }
