From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 5 May 2024 01:31:47 +0900
Subject: [PATCH] Thread-safe NeigbhorUpdater


diff --git a/src/main/java/com/vulpeslab/confetti/threading/PerThreadNeighborUpdater.java b/src/main/java/com/vulpeslab/confetti/threading/PerThreadNeighborUpdater.java
new file mode 100644
index 0000000000000000000000000000000000000000..2161635f3eacad16d4402359da4bdf0e079157c2
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/threading/PerThreadNeighborUpdater.java
@@ -0,0 +1,44 @@
+package com.vulpeslab.confetti.threading;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.redstone.NeighborUpdater;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Supplier;
+
+public class PerThreadNeighborUpdater implements NeighborUpdater {
+
+    private final ThreadLocal<NeighborUpdater> threadLocalNeighborUpdater;
+
+    public PerThreadNeighborUpdater(Supplier<NeighborUpdater> factory) {
+        this.threadLocalNeighborUpdater = ThreadLocal.withInitial(factory);
+    }
+
+    private NeighborUpdater getOrCreate() {
+        return threadLocalNeighborUpdater.get();
+    }
+
+    @Override
+    public void shapeUpdate(@NotNull Direction direction, @NotNull BlockState neighborState, @NotNull BlockPos pos, @NotNull BlockPos neighborPos, int flags, int maxUpdateDepth) {
+        getOrCreate().shapeUpdate(direction, neighborState, pos, neighborPos, flags, maxUpdateDepth);
+    }
+
+    @Override
+    public void neighborChanged(@NotNull BlockPos pos, @NotNull Block sourceBlock, @NotNull BlockPos sourcePos) {
+        getOrCreate().neighborChanged(pos, sourceBlock, sourcePos);
+    }
+
+    @Override
+    public void neighborChanged(@NotNull BlockState state, @NotNull BlockPos pos, @NotNull Block sourceBlock, @NotNull BlockPos sourcePos, boolean notify) {
+        getOrCreate().neighborChanged(state, pos, sourceBlock, sourcePos, notify);
+    }
+
+    @Override
+    public void updateNeighborsAtExceptFromFacing(@NotNull BlockPos pos, @NotNull Block sourceBlock, @Nullable Direction except) {
+        getOrCreate().updateNeighborsAtExceptFromFacing(pos, sourceBlock, except);
+    }
+}
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index ed313b0424b2e1e1335ee3e3bd4d9cfd15f24b39..15768105efeebff01c0b5de18fcb1b02a1ea7aa6 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -101,6 +101,7 @@ import org.bukkit.craftbukkit.block.data.CraftBlockData;
 import org.bukkit.craftbukkit.util.CraftSpawnCategory;
 import org.bukkit.entity.SpawnCategory;
 import org.bukkit.event.block.BlockPhysicsEvent;
+import com.vulpeslab.confetti.threading.PerThreadNeighborUpdater;
 // CraftBukkit end
 
 public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel, ca.spottedleaf.moonrise.patches.chunk_system.world.ChunkSystemEntityGetter, ca.spottedleaf.moonrise.patches.collisions.world.CollisionLevel { // Paper - rewrite chunk system // Paper - optimise collisions
@@ -770,7 +771,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         this.thread = Thread.currentThread();
         this.biomeManager = new BiomeManager(this, i);
         this.isDebug = flag1;
-        this.neighborUpdater = new CollectingNeighborUpdater(this, j);
+        this.neighborUpdater = new PerThreadNeighborUpdater(() -> new CollectingNeighborUpdater(this, j)); // Confetti - thread-safe
         this.registryAccess = iregistrycustom;
         this.damageSources = new DamageSources(iregistrycustom);
         // CraftBukkit start
