From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 21 May 2024 23:54:30 +0900
Subject: [PATCH] mpmap debug command


diff --git a/src/main/java/com/vulpeslab/confetti/commands/ConfettiCommands.java b/src/main/java/com/vulpeslab/confetti/commands/ConfettiCommands.java
new file mode 100644
index 0000000000000000000000000000000000000000..90f77c2d59bfe87ea4b23ab231a7c949ed4e392d
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/commands/ConfettiCommands.java
@@ -0,0 +1,26 @@
+package com.vulpeslab.confetti.commands;
+
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.command.Command;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ConfettiCommands {
+
+    private static final Map<String, Command> COMMANDS = new HashMap<>();
+    static {
+        for (Command command : new Command[] {
+                new MPMapCommand("mpmap")
+        }) {
+            COMMANDS.put(command.getName(), command);
+        }
+    }
+
+    public static void registerCommands(final MinecraftServer server) {
+        COMMANDS.forEach((s, command) -> {
+            server.server.getCommandMap().register(s, "confetti", command);
+        });
+    }
+
+}
diff --git a/src/main/java/com/vulpeslab/confetti/commands/MPMapCommand.java b/src/main/java/com/vulpeslab/confetti/commands/MPMapCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..eb3415e98063d67987bbf4656aa7a2749cbb7dcb
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/commands/MPMapCommand.java
@@ -0,0 +1,48 @@
+package com.vulpeslab.confetti.commands;
+
+import io.papermc.paper.chunk.system.scheduling.NewChunkHolder;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.minecraft.server.level.FullChunkStatus;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.ChunkPos;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.entity.Player;
+import com.vulpeslab.confetti.Confetti;
+
+public class MPMapCommand extends MapCommandBase {
+
+    public MPMapCommand(String command) {
+        super(command);
+        setPermission("confetti.command.mpmap");
+    }
+
+    @Override
+    protected ChunkStatus getStatus(Player player, ChunkPos chunkPos) {
+        NewChunkHolder newChunkHolder = ((ServerLevel) ((CraftPlayer) player).getHandle().level()).chunkTaskScheduler.chunkHolderManager.getChunkHolder(chunkPos.x, chunkPos.z);
+
+        NamedTextColor color;
+        String name;
+
+        if (newChunkHolder == null) {
+            color = NamedTextColor.DARK_GRAY;
+            name = "Unloaded";
+        } else if (newChunkHolder.getChunkStatus() == FullChunkStatus.INACCESSIBLE) {
+            color = NamedTextColor.GRAY;
+            name = "Inaccessible";
+        } else if (newChunkHolder.getChunkStatus() == FullChunkStatus.FULL) {
+            color = NamedTextColor.WHITE;
+            name = "Full";
+        } else if (newChunkHolder.getChunkStatus() == FullChunkStatus.BLOCK_TICKING) {
+            color = NamedTextColor.BLUE;
+            name = "Block Ticking";
+        } else if (newChunkHolder.getChunkStatus() == FullChunkStatus.ENTITY_TICKING) {
+            color = NamedTextColor.AQUA;
+            name = "Entity Ticking";
+        } else {
+            color = NamedTextColor.RED;
+            name = "Unknown";
+        }
+
+        return new ChunkStatus(color, name);
+    }
+}
diff --git a/src/main/java/com/vulpeslab/confetti/commands/MapCommandBase.java b/src/main/java/com/vulpeslab/confetti/commands/MapCommandBase.java
new file mode 100644
index 0000000000000000000000000000000000000000..a51b93baff01147de517c84e74916ab8e57f4af9
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/commands/MapCommandBase.java
@@ -0,0 +1,93 @@
+package com.vulpeslab.confetti.commands;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.event.HoverEvent;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.minecraft.world.level.ChunkPos;
+import org.bukkit.ChatColor;
+import org.bukkit.command.Command;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.entity.Player;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+public abstract class MapCommandBase extends Command {
+
+    public MapCommandBase(String command) {
+        super(command);
+    }
+
+    protected abstract ChunkStatus getStatus(Player player, ChunkPos chunkPos);
+
+    @Override
+    public boolean execute(@NotNull CommandSender sender, @NotNull String commandLabel, String[] args) {
+        if (!testPermission(sender)) return false;
+
+        if (!(sender instanceof Player player)) {
+            sender.sendMessage(ChatColor.RED + "Only players can execute this command.");
+            return false;
+        }
+
+        sendMap(player);
+
+        return true;
+    }
+
+    protected void sendMap(Player player) {
+        sendMap(player, 8);
+    }
+
+    protected void sendMap(Player player, int radius) {
+        sendHeader(player, radius);
+        for (int row = -radius; row <= radius; row++) {
+            sendRow(player, radius, row);
+        }
+        sendHeader(player, radius);
+    }
+
+    private ChunkStatus getStatus(Player player, int x, int z) {
+        return getStatus(player, new ChunkPos(((CraftPlayer) player).getHandle().blockPosition().offset(x << 4, 0, z << 4)));
+    }
+
+    private void sendRow(Player player, int radius, int row) {
+        Component component = Component.text(" | ").color(NamedTextColor.GOLD);
+
+        int i = -radius;
+        while (i <= radius) {
+            StringBuilder builder = new StringBuilder();
+            ChunkStatus status = getStatus(player, i, row);
+
+            while (i <= radius) {
+                ChunkStatus status2 = getStatus(player, i, row);
+
+                if (!Objects.equals(status, status2)) {
+                    break;
+                }
+
+                builder.append(row == 0 && i == 0 ? "\u25A0 " : "+ ");
+
+                i++;
+            }
+
+            Component innerComponent = Component.text(builder.toString()).color(status.color());
+
+            if (status.description() != null) {
+                innerComponent = innerComponent.hoverEvent(HoverEvent.showText(Component.text(status.description()).color(status.color())));
+            }
+
+            component = component.append(innerComponent);
+        }
+
+        component = component.append(Component.text("| ").color(NamedTextColor.GOLD));
+        player.sendMessage(component);
+    }
+
+    private void sendHeader(Player player, int radius) {
+        player.sendMessage(Component.text("+ " + "- ".repeat(radius * 2 + 1) + "+").color(NamedTextColor.GOLD));
+    }
+
+    protected record ChunkStatus(NamedTextColor color, @Nullable String description) {};
+}
diff --git a/src/main/java/com/vulpeslab/confetti/permissions/ConfettiCommandPermissions.java b/src/main/java/com/vulpeslab/confetti/permissions/ConfettiCommandPermissions.java
index 7ab9920729daf2cc82909d539c573028503b23ea..89a51764220ff931963d06bac58700cf8b94279b 100644
--- a/src/main/java/com/vulpeslab/confetti/permissions/ConfettiCommandPermissions.java
+++ b/src/main/java/com/vulpeslab/confetti/permissions/ConfettiCommandPermissions.java
@@ -1,6 +1,7 @@
 package com.vulpeslab.confetti.permissions;
 
 import org.bukkit.permissions.Permission;
+import org.bukkit.permissions.PermissionDefault;
 import org.bukkit.util.permissions.DefaultPermissions;
 import org.jetbrains.annotations.NotNull;
 
@@ -11,6 +12,7 @@ public class ConfettiCommandPermissions {
     public static void registerPermissions(@NotNull Permission parent) {
         Permission commands = DefaultPermissions.registerPermission(ROOT, "Gives the user the ability to use all Confetti commands", parent);
 
+        DefaultPermissions.registerPermission(PREFIX + "mpmap", "MPMap command", PermissionDefault.TRUE, commands);
 
         commands.recalculatePermissibles();
     }
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 583246cc99cd9d90f9a4205f45c2a2d5b33e096c..2bb5326e52049d02171e01ab4cf6f7fa118fd6c6 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -71,6 +71,7 @@ import org.bukkit.craftbukkit.util.TerminalConsoleWriterThread;
 import org.bukkit.event.server.ServerCommandEvent;
 import org.bukkit.craftbukkit.util.Waitable; // Paper
 import org.bukkit.event.server.RemoteServerCommandEvent;
+import com.vulpeslab.confetti.commands.ConfettiCommands;
 // CraftBukkit end
 
 public class DedicatedServer extends MinecraftServer implements ServerInterface {
@@ -248,6 +249,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
         org.purpurmc.purpur.PurpurConfig.registerCommands();
         // Purpur end
+        ConfettiCommands.registerCommands(this); // Confetti
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
         gg.pufferfish.pufferfish.PufferfishConfig.load(); // Pufferfish
         gg.pufferfish.pufferfish.PufferfishCommand.init(); // Pufferfish
