From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 3 Jul 2024 19:10:59 +0900
Subject: [PATCH] Ensure worlds are loaded on the global scheduler


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 33808b1b21c656672c5e5a57ef40668cf44ebf04..9c871774fc65789bc41ea9d03545912a19d05f3b 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -885,6 +885,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     // CraftBukkit start
     public void prepareLevels(ChunkProgressListener worldloadlistener, ServerLevel worldserver) {
         // WorldServer worldserver = this.overworld();
+        if (!TickThread.isTickThread()) throw new IllegalStateException("Cannot load worlds async!!"); // Confetti
+        if (ConfettiTickThread.isConfettiTickThread()) throw new IllegalStateException("Must load worlds from the global scheduler!"); // Confetti
         this.forceTicks = true;
         // CraftBukkit end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index c72f928c8fa5533800505488bae8284f7abc794f..8149cb89f7b715364fa67211ec6465f1f49a1fed 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1289,6 +1289,8 @@ public final class CraftServer implements Server {
         Preconditions.checkState(this.console.getAllLevels().iterator().hasNext(), "Cannot create additional worlds on STARTUP");
         //Preconditions.checkState(!this.console.isIteratingOverLevels, "Cannot create a world while worlds are being ticked"); // Paper - Cat - Temp disable. We'll see how this goes.
         Preconditions.checkArgument(creator != null, "WorldCreator cannot be null");
+        if (!TickThread.isTickThread()) throw new IllegalStateException("Cannot load worlds async!!"); // Confetti
+        if (ConfettiTickThread.isConfettiTickThread()) throw new IllegalStateException("Must load worlds from the global scheduler!"); // Confetti
 
         String name = creator.name();
         ChunkGenerator generator = creator.generator();
