From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 3 Jul 2024 19:10:59 +0900
Subject: [PATCH] Ensure worlds are loaded on the global scheduler


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 2da3f80170e1303c04b6bf72fa3df8ef4480d855..69b9de9bf12bc4a8280a164adc98313eefcb1b23 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -885,6 +885,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     // CraftBukkit start
     public void prepareLevels(ChunkProgressListener worldloadlistener, ServerLevel worldserver) {
         // WorldServer worldserver = this.overworld();
+        if (!TickThread.isTickThread()) throw new IllegalStateException("Cannot load worlds async!!"); // Confetti
+        if (ConfettiTickThread.isConfettiTickThread()) throw new IllegalStateException("Must load worlds from the global scheduler!"); // Confetti
         this.forceTicks = true;
         // CraftBukkit end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 0218475c1607d620c778ad2f0c1997e536d7f751..b5dd59ae55b9f0584bae0c7e3959b3754ff85c7f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1291,6 +1291,8 @@ public final class CraftServer implements Server {
         Preconditions.checkState(this.console.getAllLevels().iterator().hasNext(), "Cannot create additional worlds on STARTUP");
         //Preconditions.checkState(!this.console.isIteratingOverLevels, "Cannot create a world while worlds are being ticked"); // Paper - Cat - Temp disable. We'll see how this goes.
         Preconditions.checkArgument(creator != null, "WorldCreator cannot be null");
+        if (!TickThread.isTickThread()) throw new IllegalStateException("Cannot load worlds async!!"); // Confetti
+        if (ConfettiTickThread.isConfettiTickThread()) throw new IllegalStateException("Must load worlds from the global scheduler!"); // Confetti
 
         String name = creator.name();
         ChunkGenerator generator = creator.generator();
