From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: UnlimitedBytes <admin@unlimitedbytes.ovh>
Date: Thu, 1 Jan 2026 17:57:34 +0100
Subject: [PATCH] Import PlayerTeam and Scoreboard from mc-dev


diff --git a/src/main/java/net/minecraft/world/scores/PlayerTeam.java b/src/main/java/net/minecraft/world/scores/PlayerTeam.java
new file mode 100644
index 0000000000000000000000000000000000000000..9464054912e19fc78dd965b71fce20a18564b351
--- /dev/null
+++ b/src/main/java/net/minecraft/world/scores/PlayerTeam.java
@@ -0,0 +1,186 @@
+package net.minecraft.world.scores;
+
+import com.google.common.collect.Sets;
+import java.util.Collection;
+import java.util.Set;
+import javax.annotation.Nullable;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.CommonComponents;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.ComponentUtils;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.Style;
+
+public class PlayerTeam extends Team {
+    private static final int BIT_FRIENDLY_FIRE = 0;
+    private static final int BIT_SEE_INVISIBLES = 1;
+    private final Scoreboard scoreboard;
+    private final String name;
+    private final Set<String> players = Sets.newHashSet();
+    private Component displayName;
+    private Component playerPrefix = CommonComponents.EMPTY;
+    private Component playerSuffix = CommonComponents.EMPTY;
+    private boolean allowFriendlyFire = true;
+    private boolean seeFriendlyInvisibles = true;
+    private Team.Visibility nameTagVisibility = Team.Visibility.ALWAYS;
+    private Team.Visibility deathMessageVisibility = Team.Visibility.ALWAYS;
+    private ChatFormatting color = ChatFormatting.RESET;
+    private Team.CollisionRule collisionRule = Team.CollisionRule.ALWAYS;
+    private final Style displayNameStyle;
+
+    public PlayerTeam(Scoreboard scoreboard, String name) {
+        this.scoreboard = scoreboard;
+        this.name = name;
+        this.displayName = Component.literal(name);
+        this.displayNameStyle = Style.EMPTY.withInsertion(name).withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, Component.literal(name)));
+    }
+
+    public Scoreboard getScoreboard() {
+        return this.scoreboard;
+    }
+
+    @Override
+    public String getName() {
+        return this.name;
+    }
+
+    public Component getDisplayName() {
+        return this.displayName;
+    }
+
+    public MutableComponent getFormattedDisplayName() {
+        MutableComponent mutableComponent = ComponentUtils.wrapInSquareBrackets(this.displayName.copy().withStyle(this.displayNameStyle));
+        ChatFormatting chatFormatting = this.getColor();
+        if (chatFormatting != ChatFormatting.RESET) {
+            mutableComponent.withStyle(chatFormatting);
+        }
+
+        return mutableComponent;
+    }
+
+    public void setDisplayName(Component displayName) {
+        if (displayName == null) {
+            throw new IllegalArgumentException("Name cannot be null");
+        } else {
+            this.displayName = displayName;
+            this.scoreboard.onTeamChanged(this);
+        }
+    }
+
+    public void setPlayerPrefix(@Nullable Component prefix) {
+        this.playerPrefix = prefix == null ? CommonComponents.EMPTY : prefix;
+        this.scoreboard.onTeamChanged(this);
+    }
+
+    public Component getPlayerPrefix() {
+        return this.playerPrefix;
+    }
+
+    public void setPlayerSuffix(@Nullable Component suffix) {
+        this.playerSuffix = suffix == null ? CommonComponents.EMPTY : suffix;
+        this.scoreboard.onTeamChanged(this);
+    }
+
+    public Component getPlayerSuffix() {
+        return this.playerSuffix;
+    }
+
+    @Override
+    public Collection<String> getPlayers() {
+        return this.players;
+    }
+
+    @Override
+    public MutableComponent getFormattedName(Component name) {
+        MutableComponent mutableComponent = Component.empty().append(this.playerPrefix).append(name).append(this.playerSuffix);
+        ChatFormatting chatFormatting = this.getColor();
+        if (chatFormatting != ChatFormatting.RESET) {
+            mutableComponent.withStyle(chatFormatting);
+        }
+
+        return mutableComponent;
+    }
+
+    public static MutableComponent formatNameForTeam(@Nullable Team team, Component name) {
+        return team == null ? name.copy() : team.getFormattedName(name);
+    }
+
+    @Override
+    public boolean isAllowFriendlyFire() {
+        return this.allowFriendlyFire;
+    }
+
+    public void setAllowFriendlyFire(boolean friendlyFire) {
+        this.allowFriendlyFire = friendlyFire;
+        this.scoreboard.onTeamChanged(this);
+    }
+
+    @Override
+    public boolean canSeeFriendlyInvisibles() {
+        return this.seeFriendlyInvisibles;
+    }
+
+    public void setSeeFriendlyInvisibles(boolean showFriendlyInvisible) {
+        this.seeFriendlyInvisibles = showFriendlyInvisible;
+        this.scoreboard.onTeamChanged(this);
+    }
+
+    @Override
+    public Team.Visibility getNameTagVisibility() {
+        return this.nameTagVisibility;
+    }
+
+    @Override
+    public Team.Visibility getDeathMessageVisibility() {
+        return this.deathMessageVisibility;
+    }
+
+    public void setNameTagVisibility(Team.Visibility nameTagVisibilityRule) {
+        this.nameTagVisibility = nameTagVisibilityRule;
+        this.scoreboard.onTeamChanged(this);
+    }
+
+    public void setDeathMessageVisibility(Team.Visibility deathMessageVisibilityRule) {
+        this.deathMessageVisibility = deathMessageVisibilityRule;
+        this.scoreboard.onTeamChanged(this);
+    }
+
+    @Override
+    public Team.CollisionRule getCollisionRule() {
+        return this.collisionRule;
+    }
+
+    public void setCollisionRule(Team.CollisionRule collisionRule) {
+        this.collisionRule = collisionRule;
+        this.scoreboard.onTeamChanged(this);
+    }
+
+    public int packOptions() {
+        int i = 0;
+        if (this.isAllowFriendlyFire()) {
+            i |= 1;
+        }
+
+        if (this.canSeeFriendlyInvisibles()) {
+            i |= 2;
+        }
+
+        return i;
+    }
+
+    public void unpackOptions(int flags) {
+        this.setAllowFriendlyFire((flags & 1) > 0);
+        this.setSeeFriendlyInvisibles((flags & 2) > 0);
+    }
+
+    public void setColor(ChatFormatting color) {
+        this.color = color;
+        this.scoreboard.onTeamChanged(this);
+    }
+
+    @Override
+    public ChatFormatting getColor() {
+        return this.color;
+    }
+}
diff --git a/src/main/java/net/minecraft/world/scores/Scoreboard.java b/src/main/java/net/minecraft/world/scores/Scoreboard.java
new file mode 100644
index 0000000000000000000000000000000000000000..121b57ec018bffc05904e7596e0963e2bb545dc1
--- /dev/null
+++ b/src/main/java/net/minecraft/world/scores/Scoreboard.java
@@ -0,0 +1,383 @@
+package net.minecraft.world.scores;
+
+import com.google.common.collect.Lists;
+import com.mojang.logging.LogUtils;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntMaps;
+import it.unimi.dsi.fastutil.objects.Object2ObjectMap;
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.Reference2ObjectMap;
+import it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+import net.minecraft.core.HolderLookup;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.ListTag;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.numbers.NumberFormat;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.scores.criteria.ObjectiveCriteria;
+import org.apache.commons.lang3.mutable.MutableBoolean;
+import org.slf4j.Logger;
+
+public class Scoreboard {
+    public static final String HIDDEN_SCORE_PREFIX = "#";
+    private static final Logger LOGGER = LogUtils.getLogger();
+    private final Object2ObjectMap<String, Objective> objectivesByName = new Object2ObjectOpenHashMap<>(16, 0.5F);
+    private final Reference2ObjectMap<ObjectiveCriteria, List<Objective>> objectivesByCriteria = new Reference2ObjectOpenHashMap<>();
+    private final Map<String, PlayerScores> playerScores = new Object2ObjectOpenHashMap<>(16, 0.5F);
+    private final Map<DisplaySlot, Objective> displayObjectives = new EnumMap<>(DisplaySlot.class);
+    private final Object2ObjectMap<String, PlayerTeam> teamsByName = new Object2ObjectOpenHashMap<>();
+    private final Object2ObjectMap<String, PlayerTeam> teamsByPlayer = new Object2ObjectOpenHashMap<>();
+
+    @Nullable
+    public Objective getObjective(@Nullable String name) {
+        return this.objectivesByName.get(name);
+    }
+
+    public Objective addObjective(
+        String name,
+        ObjectiveCriteria criterion,
+        Component displayName,
+        ObjectiveCriteria.RenderType renderType,
+        boolean displayAutoUpdate,
+        @Nullable NumberFormat numberFormat
+    ) {
+        if (this.objectivesByName.containsKey(name)) {
+            throw new IllegalArgumentException("An objective with the name '" + name + "' already exists!");
+        } else {
+            Objective objective = new Objective(this, name, criterion, displayName, renderType, displayAutoUpdate, numberFormat);
+            this.objectivesByCriteria.computeIfAbsent(criterion, criterion2 -> Lists.newArrayList()).add(objective);
+            this.objectivesByName.put(name, objective);
+            this.onObjectiveAdded(objective);
+            return objective;
+        }
+    }
+
+    public final void forAllObjectives(ObjectiveCriteria criterion, ScoreHolder scoreHolder, Consumer<ScoreAccess> action) {
+        this.objectivesByCriteria
+            .getOrDefault(criterion, Collections.emptyList())
+            .forEach(objective -> action.accept(this.getOrCreatePlayerScore(scoreHolder, objective, true)));
+    }
+
+    private PlayerScores getOrCreatePlayerInfo(String scoreHolderName) {
+        return this.playerScores.computeIfAbsent(scoreHolderName, name -> new PlayerScores());
+    }
+
+    public ScoreAccess getOrCreatePlayerScore(ScoreHolder scoreHolder, Objective objective) {
+        return this.getOrCreatePlayerScore(scoreHolder, objective, false);
+    }
+
+    public ScoreAccess getOrCreatePlayerScore(ScoreHolder scoreHolder, Objective objective, boolean forceWritable) {
+        final boolean bl = forceWritable || !objective.getCriteria().isReadOnly();
+        PlayerScores playerScores = this.getOrCreatePlayerInfo(scoreHolder.getScoreboardName());
+        final MutableBoolean mutableBoolean = new MutableBoolean();
+        final Score score = playerScores.getOrCreate(objective, scorex -> mutableBoolean.setTrue());
+        return new ScoreAccess() {
+            @Override
+            public int get() {
+                return score.value();
+            }
+
+            @Override
+            public void set(int score) {
+                if (!bl) {
+                    throw new IllegalStateException("Cannot modify read-only score");
+                } else {
+                    boolean bl = mutableBoolean.isTrue();
+                    if (objective.displayAutoUpdate()) {
+                        Component component = scoreHolder.getDisplayName();
+                        if (component != null && !component.equals(score.display())) {
+                            score.display(component);
+                            bl = true;
+                        }
+                    }
+
+                    if (score != score.value()) {
+                        score.value(score);
+                        bl = true;
+                    }
+
+                    if (bl) {
+                        this.sendScoreToPlayers();
+                    }
+                }
+            }
+
+            @Nullable
+            @Override
+            public Component display() {
+                return score.display();
+            }
+
+            @Override
+            public void display(@Nullable Component text) {
+                if (mutableBoolean.isTrue() || !Objects.equals(text, score.display())) {
+                    score.display(text);
+                    this.sendScoreToPlayers();
+                }
+            }
+
+            @Override
+            public void numberFormatOverride(@Nullable NumberFormat numberFormat) {
+                score.numberFormat(numberFormat);
+                this.sendScoreToPlayers();
+            }
+
+            @Override
+            public boolean locked() {
+                return score.isLocked();
+            }
+
+            @Override
+            public void unlock() {
+                this.setLocked(false);
+            }
+
+            @Override
+            public void lock() {
+                this.setLocked(true);
+            }
+
+            private void setLocked(boolean locked) {
+                score.setLocked(locked);
+                if (mutableBoolean.isTrue()) {
+                    this.sendScoreToPlayers();
+                }
+
+                Scoreboard.this.onScoreLockChanged(scoreHolder, objective);
+            }
+
+            private void sendScoreToPlayers() {
+                Scoreboard.this.onScoreChanged(scoreHolder, objective, score);
+                mutableBoolean.setFalse();
+            }
+        };
+    }
+
+    @Nullable
+    public ReadOnlyScoreInfo getPlayerScoreInfo(ScoreHolder scoreHolder, Objective objective) {
+        PlayerScores playerScores = this.playerScores.get(scoreHolder.getScoreboardName());
+        return playerScores != null ? playerScores.get(objective) : null;
+    }
+
+    public Collection<PlayerScoreEntry> listPlayerScores(Objective objective) {
+        List<PlayerScoreEntry> list = new ArrayList<>();
+        this.playerScores.forEach((scoreHolderName, scores) -> {
+            Score score = scores.get(objective);
+            if (score != null) {
+                list.add(new PlayerScoreEntry(scoreHolderName, score.value(), score.display(), score.numberFormat()));
+            }
+        });
+        return list;
+    }
+
+    public Collection<Objective> getObjectives() {
+        return this.objectivesByName.values();
+    }
+
+    public Collection<String> getObjectiveNames() {
+        return this.objectivesByName.keySet();
+    }
+
+    public Collection<ScoreHolder> getTrackedPlayers() {
+        return this.playerScores.keySet().stream().map(ScoreHolder::forNameOnly).toList();
+    }
+
+    public void resetAllPlayerScores(ScoreHolder scoreHolder) {
+        PlayerScores playerScores = this.playerScores.remove(scoreHolder.getScoreboardName());
+        if (playerScores != null) {
+            this.onPlayerRemoved(scoreHolder);
+        }
+    }
+
+    public void resetSinglePlayerScore(ScoreHolder scoreHolder, Objective objective) {
+        PlayerScores playerScores = this.playerScores.get(scoreHolder.getScoreboardName());
+        if (playerScores != null) {
+            boolean bl = playerScores.remove(objective);
+            if (!playerScores.hasScores()) {
+                PlayerScores playerScores2 = this.playerScores.remove(scoreHolder.getScoreboardName());
+                if (playerScores2 != null) {
+                    this.onPlayerRemoved(scoreHolder);
+                }
+            } else if (bl) {
+                this.onPlayerScoreRemoved(scoreHolder, objective);
+            }
+        }
+    }
+
+    public Object2IntMap<Objective> listPlayerScores(ScoreHolder scoreHolder) {
+        PlayerScores playerScores = this.playerScores.get(scoreHolder.getScoreboardName());
+        return playerScores != null ? playerScores.listScores() : Object2IntMaps.emptyMap();
+    }
+
+    public void removeObjective(Objective objective) {
+        this.objectivesByName.remove(objective.getName());
+
+        for (DisplaySlot displaySlot : DisplaySlot.values()) {
+            if (this.getDisplayObjective(displaySlot) == objective) {
+                this.setDisplayObjective(displaySlot, null);
+            }
+        }
+
+        List<Objective> list = this.objectivesByCriteria.get(objective.getCriteria());
+        if (list != null) {
+            list.remove(objective);
+        }
+
+        for (PlayerScores playerScores : this.playerScores.values()) {
+            playerScores.remove(objective);
+        }
+
+        this.onObjectiveRemoved(objective);
+    }
+
+    public void setDisplayObjective(DisplaySlot slot, @Nullable Objective objective) {
+        this.displayObjectives.put(slot, objective);
+    }
+
+    @Nullable
+    public Objective getDisplayObjective(DisplaySlot slot) {
+        return this.displayObjectives.get(slot);
+    }
+
+    @Nullable
+    public PlayerTeam getPlayerTeam(String name) {
+        return this.teamsByName.get(name);
+    }
+
+    public PlayerTeam addPlayerTeam(String name) {
+        PlayerTeam playerTeam = this.getPlayerTeam(name);
+        if (playerTeam != null) {
+            LOGGER.warn("Requested creation of existing team '{}'", name);
+            return playerTeam;
+        } else {
+            playerTeam = new PlayerTeam(this, name);
+            this.teamsByName.put(name, playerTeam);
+            this.onTeamAdded(playerTeam);
+            return playerTeam;
+        }
+    }
+
+    public void removePlayerTeam(PlayerTeam team) {
+        this.teamsByName.remove(team.getName());
+
+        for (String string : team.getPlayers()) {
+            this.teamsByPlayer.remove(string);
+        }
+
+        this.onTeamRemoved(team);
+    }
+
+    public boolean addPlayerToTeam(String scoreHolderName, PlayerTeam team) {
+        if (this.getPlayersTeam(scoreHolderName) != null) {
+            this.removePlayerFromTeam(scoreHolderName);
+        }
+
+        this.teamsByPlayer.put(scoreHolderName, team);
+        return team.getPlayers().add(scoreHolderName);
+    }
+
+    public boolean removePlayerFromTeam(String scoreHolderName) {
+        PlayerTeam playerTeam = this.getPlayersTeam(scoreHolderName);
+        if (playerTeam != null) {
+            this.removePlayerFromTeam(scoreHolderName, playerTeam);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public void removePlayerFromTeam(String scoreHolderName, PlayerTeam team) {
+        if (this.getPlayersTeam(scoreHolderName) != team) {
+            throw new IllegalStateException("Player is either on another team or not on any team. Cannot remove from team '" + team.getName() + "'.");
+        } else {
+            this.teamsByPlayer.remove(scoreHolderName);
+            team.getPlayers().remove(scoreHolderName);
+        }
+    }
+
+    public Collection<String> getTeamNames() {
+        return this.teamsByName.keySet();
+    }
+
+    public Collection<PlayerTeam> getPlayerTeams() {
+        return this.teamsByName.values();
+    }
+
+    @Nullable
+    public PlayerTeam getPlayersTeam(String scoreHolderName) {
+        return this.teamsByPlayer.get(scoreHolderName);
+    }
+
+    public void onObjectiveAdded(Objective objective) {
+    }
+
+    public void onObjectiveChanged(Objective objective) {
+    }
+
+    public void onObjectiveRemoved(Objective objective) {
+    }
+
+    protected void onScoreChanged(ScoreHolder scoreHolder, Objective objective, Score score) {
+    }
+
+    protected void onScoreLockChanged(ScoreHolder scoreHolder, Objective objective) {
+    }
+
+    public void onPlayerRemoved(ScoreHolder scoreHolder) {
+    }
+
+    public void onPlayerScoreRemoved(ScoreHolder scoreHolder, Objective objective) {
+    }
+
+    public void onTeamAdded(PlayerTeam team) {
+    }
+
+    public void onTeamChanged(PlayerTeam team) {
+    }
+
+    public void onTeamRemoved(PlayerTeam team) {
+    }
+
+    public void entityRemoved(Entity entity) {
+        if (!(entity instanceof Player) && !entity.isAlive()) {
+            this.resetAllPlayerScores(entity);
+            this.removePlayerFromTeam(entity.getScoreboardName());
+        }
+    }
+
+    protected ListTag savePlayerScores(HolderLookup.Provider registries) {
+        ListTag listTag = new ListTag();
+        this.playerScores.forEach((name, scores) -> scores.listRawScores().forEach((objective, score) -> {
+                CompoundTag compoundTag = score.write(registries);
+                compoundTag.putString("Name", name);
+                compoundTag.putString("Objective", objective.getName());
+                listTag.add(compoundTag);
+            }));
+        return listTag;
+    }
+
+    protected void loadPlayerScores(ListTag list, HolderLookup.Provider registries) {
+        for (int i = 0; i < list.size(); i++) {
+            CompoundTag compoundTag = list.getCompound(i);
+            Score score = Score.read(compoundTag, registries);
+            String string = compoundTag.getString("Name");
+            String string2 = compoundTag.getString("Objective");
+            Objective objective = this.getObjective(string2);
+            if (objective == null) {
+                LOGGER.error("Unknown objective {} for name {}, ignoring", string2, string);
+            } else {
+                this.getOrCreatePlayerInfo(string).setScore(objective, score);
+            }
+        }
+    }
+}
