From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 23 Aug 2024 01:03:30 +0900
Subject: [PATCH] Moving into another region


diff --git a/src/main/java/com/vulpeslab/confetti/region/RegionPos.java b/src/main/java/com/vulpeslab/confetti/region/RegionPos.java
index 64231d4c3fcee32f8243cc0d7c75795c89b55a78..a6b574df85a04147136d81ee3537b8676f4c5e9e 100644
--- a/src/main/java/com/vulpeslab/confetti/region/RegionPos.java
+++ b/src/main/java/com/vulpeslab/confetti/region/RegionPos.java
@@ -16,6 +16,7 @@ public class RegionPos {
     public static final int REGION_SIZE; // eg 8 (for an 8x8 region)
     public static final int REGION_SHIFT; // eg 3 (1 << 3 == 8)
     public static final int REGION_SIZE_MASK; // eg 7 (9 % 8 == 9 & 7 == 1)
+    public static final int MAX_DISTANCE_SQR;
 
     static {
         // desiredRegionSize = 7 -> shift = 3, size = 8, mask = 7
@@ -43,6 +44,8 @@ public class RegionPos {
         }
 
         LOGGER.info("Using region size: {}, shift={}, mask={}", REGION_SIZE, REGION_SHIFT, REGION_SIZE_MASK);
+
+        MAX_DISTANCE_SQR = RegionPos.REGION_SIZE * 16 * RegionPos.REGION_SIZE * 16;
     }
 
     public final int x;
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 4defb73aff9907f884320af039a509084e3eb0a3..689c9832b3ef911ecc1f08e2ba266e7b791a570e 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1141,7 +1141,9 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public void move(MoverType movementType, Vec3 movement) {
         final Vec3 originalMovement = movement; // Paper - Expose pre-collision velocity
         // Paper start - detailed watchdog information
-        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread("Cannot move an entity off-main");
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot move an entity off-main"); // Confetti - check source region
+        BlockPos newPos = new BlockPos((int) (this.getX() + movement.x), (int) (this.getY() + movement.y), (int) (this.getZ() + movement.z)); // Confetti - check destination
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread((ServerLevel) this.level, newPos, "Cannot move an entity into off-main"); // Confetti - check destination region
         synchronized (this.posLock) {
             this.moveStartX = this.getX();
             this.moveStartY = this.getY();
@@ -2255,7 +2257,13 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             }
             delta = event.getKnockback();
         }
-        this.setDeltaMovement(this.getDeltaMovement().add(delta.getX(), delta.getY(), delta.getZ()));
+        // Confetti start - limit push velocity
+        Vec3 newDelta = this.getDeltaMovement().add(delta.getX(), delta.getY(), delta.getZ());
+        if (newDelta.lengthSqr() > RegionPos.MAX_DISTANCE_SQR) {
+            newDelta = newDelta.normalize().scale(RegionPos.REGION_SIZE * 16);
+        }
+        this.setDeltaMovement(newDelta);
+        // Confetti end - limit push velocity
         // Paper end - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
         this.hasImpulse = true;
     }
@@ -4681,6 +4689,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public void setDeltaMovement(Vec3 velocity) {
+        // Confetti start - why is something setting the entity velocity to larger than one region...
+        if (velocity.horizontalDistanceSqr() > RegionPos.MAX_DISTANCE_SQR + 1 && velocity.horizontalDistanceSqr() > this.deltaMovement.horizontalDistanceSqr()) {
+            LOGGER.warn("Velocity is being set larger than the Confetti region size: {} for entity {}", velocity, this, new Exception("Velocity larger than region size"));
+        }
+        // Confetti end - why is something setting the entity velocity to larger than one region...
         synchronized (this.posLock) { // Paper
         this.deltaMovement = velocity;
         } // Paper
diff --git a/src/main/java/net/minecraft/world/entity/projectile/EyeOfEnder.java b/src/main/java/net/minecraft/world/entity/projectile/EyeOfEnder.java
index fca3786d0a3f99a3e61e7a4b2251361276eff9d7..8862791ac91935015cc3927156ed5ad7dccd089b 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/EyeOfEnder.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/EyeOfEnder.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.entity.projectile;
 
+import com.vulpeslab.confetti.region.RegionPos;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.nbt.CompoundTag;
@@ -137,6 +138,8 @@ public class EyeOfEnder extends Entity implements ItemSupplier {
 
             int i = this.getY() < this.ty ? 1 : -1;
 
+            if (d6 > RegionPos.MAX_DISTANCE_SQR) d6 = RegionPos.MAX_DISTANCE_SQR; // Confetti - keep within a region
+
             vec3d = new Vec3(Math.cos((double) f1) * d6, d7 + ((double) i - d7) * 0.014999999664723873D, Math.sin((double) f1) * d6);
             this.setDeltaMovement(vec3d);
         }
