From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 1 May 2025 23:21:56 +0900
Subject: [PATCH] Optimization: vanish api


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 173e4a075078af67f030750c9a6294ab3f796677..8d86dbc1a75701c202e1a932258ae55826139d67 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -5,6 +5,7 @@ import com.google.common.collect.ImmutableSet;
 import com.google.common.io.BaseEncoding;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Pair;
+import com.vulpeslab.confetti.config.ConfettiConfiguration;
 import io.netty.buffer.Unpooled;
 import it.unimi.dsi.fastutil.shorts.ShortArraySet;
 import it.unimi.dsi.fastutil.shorts.ShortSet;
@@ -2254,15 +2255,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public boolean canSee(Player player) {
+        if (ConfettiConfiguration.get().optimizations.disableVanishApi) return true; // Confetti - optimize canSee
         return this.canSee((org.bukkit.entity.Entity) player);
     }
 
     @Override
     public boolean canSee(org.bukkit.entity.Entity entity) {
+        if (ConfettiConfiguration.get().optimizations.disableVanishApi) return true; // Confetti - optimize canSee
         return this.equals(entity) || entity.isVisibleByDefault() ^ this.invertedVisibilityEntities.containsKey(entity.getUniqueId()); // SPIGOT-7312: Can always see self
     }
 
     public boolean canSeePlayer(UUID uuid) {
+        if (ConfettiConfiguration.get().optimizations.disableVanishApi) return true; // Confetti - optimize canSee
         org.bukkit.entity.Entity entity = this.getServer().getPlayer(uuid);
 
         return (entity != null) ? this.canSee(entity) : false; // If we can't find it, we can't see it
