From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: UnlimitedBytes <admin@unlimitedbytes.ovh>
Date: Mon, 9 Jun 2025 01:04:56 +0200
Subject: [PATCH] Fix EntityRemoveEventTest by adding missing Bukkit remove
 cause to Projectile discard


diff --git a/src/main/java/net/minecraft/world/entity/projectile/Projectile.java b/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
index 2d4e363c08250bcaed0057857f9bffc9a34f10c8..7a77757da57c470c53a9bffd914ea3333f7c7abc 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
@@ -71,7 +71,7 @@ public abstract class Projectile extends Entity implements TraceableEntity {
             if (!isLoaded) {
                 if (Projectile.loadedThisTick > gg.pufferfish.pufferfish.PufferfishConfig.maxProjectileLoadsPerTick) {
                     if (++this.loadedLifetime > gg.pufferfish.pufferfish.PufferfishConfig.maxProjectileLoadsPerProjectile) {
-                        this.discard();
+                        this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // Confetti - add Bukkit remove cause
                     }
                     return;
                 }
