From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 16 Jul 2025 18:42:23 +0900
Subject: [PATCH] Optimization: Use lazyExecute if we aren't flushing


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index a937bc49a85c0e565f0400163c1c9174891fb48d..89293756fdea3768e3cca5801dbf46146f86027f 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -4,6 +4,7 @@ import com.google.common.base.Suppliers;
 import com.google.common.collect.Queues;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.mojang.logging.LogUtils;
+import com.vulpeslab.confetti.config.ConfettiConfiguration;
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelException;
@@ -42,6 +43,7 @@ import java.util.function.Consumer;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
 import javax.crypto.Cipher;
+import io.netty.util.concurrent.AbstractEventExecutor;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.network.chat.Component;
@@ -494,6 +496,14 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
         if (this.channel.eventLoop().inEventLoop()) {
             this.doSendPacket(packet, callbacks, flush);
         } else {
+            // Confetti start - Use lazyExecute if we aren't flushing
+            if (ConfettiConfiguration.get().optimizations.useLazyExecuteWhenNotFlushing && !flush) {
+                ((AbstractEventExecutor) this.channel.eventLoop()).lazyExecute(() -> {
+                    this.doSendPacket(packet, callbacks, flush);
+                });
+                return;
+            }
+            // Confetti end - Use lazyExecute if we aren't flushing
             this.channel.eventLoop().execute(() -> {
                 this.doSendPacket(packet, callbacks, flush);
             });
