From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 16 Jul 2025 19:19:00 +0900
Subject: [PATCH] Optimization: Don't check isConnected


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index 89293756fdea3768e3cca5801dbf46146f86027f..768b70e04af572804d66aa5bf02af36c2ff38877 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -447,7 +447,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
     public void send(Packet<?> packet, @Nullable PacketSendListener callbacks, boolean flush) {
         // Paper start - Optimize network: Handle oversized packets better
-        final boolean connected = this.isConnected();
+        final boolean connected = true || this.isConnected(); // Confetti - Checking if the player is connected is expensive
         if (!connected && !this.preparing) {
             return;
         }
