From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 18 Jul 2025 14:24:19 +0900
Subject: [PATCH] Optimization: maximum trackers per entity


diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index 573d4c7c144bb4a991fa41255d9a277f6e0ca8f9..41f6925c54a992a4320758104ac39cde692fd33d 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -10,6 +10,7 @@ import com.google.common.collect.Queues;
 import com.google.common.collect.Sets;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
+import com.vulpeslab.confetti.config.ConfettiConfiguration;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.longs.Long2ByteMap;
@@ -21,12 +22,15 @@ import it.unimi.dsi.fastutil.longs.Long2ObjectMap.Entry;
 import it.unimi.dsi.fastutil.longs.LongIterator;
 import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
 import it.unimi.dsi.fastutil.longs.LongSet;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ObjectBidirectionalIterator;
 import it.unimi.dsi.fastutil.objects.ObjectIterator;
 import java.io.IOException;
 import java.io.Writer;
 import java.nio.file.Path;
 import java.util.ArrayList;
+import java.util.Collection;
+import java.util.ConcurrentModificationException;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
@@ -36,6 +40,7 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.Queue;
 import java.util.Set;
+import java.util.TreeSet;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
@@ -48,6 +53,8 @@ import java.util.function.IntFunction;
 import java.util.function.IntSupplier;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
+
+import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
@@ -103,6 +110,7 @@ import net.minecraft.world.level.storage.DimensionDataStorage;
 import net.minecraft.world.level.storage.LevelStorageSource;
 import net.minecraft.world.phys.Vec3;
 import org.apache.commons.lang3.mutable.MutableBoolean;
+import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
 
 // CraftBukkit start
@@ -1132,13 +1140,24 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
             this.lastChunkUpdate = currChunkUpdate;
             this.lastTrackedChunk = chunk;
 
+            // Confetti start - trackerFullUpdateFrequency and maximum trackers per entity
+            boolean fullTrackerUpdate = (level.getGameTime() ^ (long) this.entity.getId()) % ConfettiConfiguration.get().optimizations.trackerFullUpdateFrequency == 0L;
+            if (fullTrackerUpdate && this.entity instanceof ServerPlayer player) player.hasMaximumTrackerBypassPermission = player.getBukkitEntity().hasPermission("confetti.maximumtrackerbypass");
+            if (players.size() > ConfettiConfiguration.get().optimizations.maximumTrackersPerEntity) {
+                updatePlayersLimitedAndOrdered(players, lastChunkUpdate != currChunkUpdate || lastTrackedChunk != chunk, fullTrackerUpdate);
+                return;
+            }
+            // Confetti end - trackerFullUpdateFrequency and maximum trackers per entity
+
             final ServerPlayer[] playersRaw = players.getRawDataUnchecked();
 
             for (int i = 0, len = players.size(); i < len; ++i) {
                 final ServerPlayer player = playersRaw[i];
-                this.updatePlayer(player);
+                if (fullTrackerUpdate || !this.seenBy.contains(player.connection)) this.updatePlayer(player); // Confetti - trackerFullUpdateFrequency
             }
 
+            if (!fullTrackerUpdate) return; // Confetti - trackerFullUpdateFrequency
+
             if (lastChunkUpdate != currChunkUpdate || lastTrackedChunk != chunk) {
                 // need to purge any players possible not in the chunk list
                 seenByLock.lock(); try { // Confetti - seenBy thread safety
@@ -1199,6 +1218,91 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
             this.lastSectionPos = SectionPos.of((EntityAccess) entity);
         }
 
+        // Confetti start - maximum trackers per entity
+        private record DistancedServerPlayer(ServerPlayer serverPlayer, double distanceSqr, boolean hasBypassPerm) implements Comparable<DistancedServerPlayer> {
+            DistancedServerPlayer(ServerPlayer serverPlayer, Entity entity) {
+                this(serverPlayer, distanceBetween(serverPlayer, entity), serverPlayer.hasMaximumTrackerBypassPermission);
+            }
+
+            private static double distanceBetween(ServerPlayer serverPlayer, Entity entity) {
+                double vec3d_dx = serverPlayer.getX() - entity.getX();
+                double vec3d_dz = serverPlayer.getZ() - entity.getZ();
+                return vec3d_dx * vec3d_dx + vec3d_dz * vec3d_dz;
+            }
+
+            @Override
+            public int compareTo(@NotNull ChunkMap.TrackedEntity.DistancedServerPlayer o) {
+                int compareResult = Double.compare(hasBypassPerm ? 0 : this.distanceSqr, o.hasBypassPerm ? 0 : o.distanceSqr);
+                return compareResult == 0 ? Integer.compare(this.serverPlayer.getId(), o.serverPlayer.getId()) : compareResult;
+            }
+        }
+
+        public void updatePlayersLimitedAndOrdered(ca.spottedleaf.moonrise.common.list.ReferenceList<ServerPlayer> players, boolean hasTrackerCandidatesChanged, boolean fullTrackerUpdate) {
+            if (!hasTrackerCandidatesChanged && !fullTrackerUpdate) return;
+
+            TreeSet<DistancedServerPlayer> playerSet = new TreeSet<>();
+            final ServerPlayer[] playersRaw = players.getRawDataUnchecked();
+
+            for (int index = 0, len = players.size(); index < len; ++index) {
+                ServerPlayer player = playersRaw[index];
+                if (player == this.entity) continue;
+                DistancedServerPlayer distancedPlayer = new DistancedServerPlayer(player, this.entity);
+                int i = ChunkMap.this.getPlayerViewDistance(player);
+                double d0 = (double) Math.min(this.getEffectiveRange(), i * 16);
+                double d2 = d0 * d0;
+                boolean flag = distancedPlayer.distanceSqr <= d2 && this.entity.broadcastToPlayer(player) && ChunkMap.this.isChunkTracked(player, this.entity.chunkPosition().x, this.entity.chunkPosition().z);
+                // Paper start - Configurable entity tracking range by Y
+                if (flag && level.paperConfig().entities.trackingRangeY.enabled) {
+                    double rangeY = level.paperConfig().entities.trackingRangeY.get(this.entity, -1);
+                    if (rangeY != -1) {
+                        double vec3d_dy = player.getY() - this.entity.getY();
+                        flag = vec3d_dy * vec3d_dy <= rangeY * rangeY;
+                    }
+                }
+                // Paper end - Configurable entity tracking range by Y
+
+                // CraftBukkit start - respect vanish API
+                if (flag && !player.getBukkitEntity().canSee(this.entity.getBukkitEntity())) { // Paper - only consider hits
+                    flag = false;
+                }
+
+                if (flag) {
+                    playerSet.add(distancedPlayer);
+                }
+            }
+
+            seenByLock.lock();
+            try {
+                Set<ServerPlayer> canBeSeenBy = new ReferenceOpenHashSet<>(Math.min(ConfettiConfiguration.get().optimizations.maximumTrackersPerEntity, playerSet.size()));
+                Iterator<DistancedServerPlayer> playerSetIterator = playerSet.iterator();
+                int count = 0;
+                while (playerSetIterator.hasNext() && count++ < ConfettiConfiguration.get().optimizations.maximumTrackersPerEntity) { // Confetti - maximum trackers per entity
+                    DistancedServerPlayer distancedPlayer = playerSetIterator.next();
+                    ServerPlayer player = distancedPlayer.serverPlayer();
+                    canBeSeenBy.add(player);
+                    if (this.seenBy.add(player.connection)) {
+                        // Paper start - entity tracking events
+                        if (io.papermc.paper.event.player.PlayerTrackEntityEvent.getHandlerList().getRegisteredListeners().length == 0 || new io.papermc.paper.event.player.PlayerTrackEntityEvent(player.getBukkitEntity(), this.entity.getBukkitEntity()).callEvent()) {
+                            this.serverEntity.addPairing(player);
+                        }
+                        // Paper end - entity tracking events
+                        this.serverEntity.onPlayerAdd(); // Paper - fix desync when a player is added to the tracker
+                    }
+                }
+
+                if (this.seenBy.size() != canBeSeenBy.size()) {
+                    for (ServerPlayerConnection conn : this.seenBy.toArray(new ServerPlayerConnection[0])) { // avoid CME
+                        if (!canBeSeenBy.contains(conn.getPlayer()) && this.seenBy.remove(conn)) {
+                            this.serverEntity.removePairing(conn.getPlayer());
+                        }
+                    }
+                }
+            } finally {
+                seenByLock.unlock();
+            }
+        }
+        // Confetti end - maximum trackers per entity
+
         public boolean equals(Object object) {
             return object instanceof ChunkMap.TrackedEntity ? ((ChunkMap.TrackedEntity) object).entity.getId() == this.entity.getId() : false;
         }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index d5878384d71c31a5bca24b55dca01d6eb5e9c776..8d2bc1c7fbfb8fa270a256972813c9b2769980e7 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -254,6 +254,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
     private BlockPos respawnPosition;
     private boolean respawnForced;
     private float respawnAngle;
+    public boolean hasMaximumTrackerBypassPermission; // Confetti
     private final TextFilter textFilter;
     private boolean textFilteringEnabled;
     private boolean allowsListing;
