From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 10 Aug 2025 18:50:42 +0900
Subject: [PATCH] BroadcastPacketEvent


diff --git a/src/main/java/com/vulpeslab/confetti/event/BroadcastPacketEvent.java b/src/main/java/com/vulpeslab/confetti/event/BroadcastPacketEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..62f7adcfb0823c19d2af727fa7da82299efb06ad
--- /dev/null
+++ b/src/main/java/com/vulpeslab/confetti/event/BroadcastPacketEvent.java
@@ -0,0 +1,67 @@
+package com.vulpeslab.confetti.event;
+
+import net.minecraft.network.protocol.Packet;
+import org.bukkit.World;
+import org.bukkit.entity.Entity;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Called whenever a packet is broadcasted to players, but will not be sent to
+ * the player that initiates the packet. Will be called even if no players will
+ * receive the packet.
+ */
+@ApiStatus.Experimental
+public class BroadcastPacketEvent extends Event {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+
+    private final World world;
+    private final Packet<?> packet;
+    private final Entity source;
+
+    public BroadcastPacketEvent(@NotNull World world, @NotNull Packet<?> packet, @Nullable Entity source) {
+        this.world = world;
+        this.packet = packet;
+        this.source = source;
+    }
+
+    /**
+     * The world the packet is being broadcasted in.
+     */
+    @NotNull
+    public World getWorld() {
+        return world;
+    }
+
+    /**
+     * The packet being broadcasted.
+     */
+    @NotNull
+    public Packet<?> getPacket() {
+        return packet;
+    }
+
+    /**
+     * Gets the entity that triggered this packet to be broadcasted. May be null
+     * if the packet was not triggered by an entity.
+     */
+    @Nullable
+    public Entity getSource() {
+        return source;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+}
diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index 41f6925c54a992a4320758104ac39cde692fd33d..34bc95e034ebd24f0a345a66852956032ddc6371 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -1322,6 +1322,8 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
             }
 
             } finally { seenByLock.unlock(); } // ShreededPaper - seenBy thread safety
+
+            new com.vulpeslab.confetti.event.BroadcastPacketEvent(ChunkMap.this.level.getWorld(), packet, this.entity.getBukkitEntity()).callEvent(); // Confetti - BroadcastPacketEvent
         }
 
         public void broadcastAndSend(Packet<?> packet) {
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 8dae6929a9194d3916f3b949907131d56b7d5e96..b2211668ce08e168e9696a61ce51bda8e61f15df 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1667,7 +1667,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
 
     @Override
     public void destroyBlockProgress(int entityId, BlockPos pos, int progress) {
-        Iterator iterator = this.server.getPlayerList().getPlayers().iterator();
+        Iterator iterator = this.players.iterator(); // Confetti - BroadcastPacketEvent - optimisation: use the level's players instead of all players
 
         // CraftBukkit start
         Player entityhuman = null;
@@ -1686,6 +1686,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         }
         // Paper end - Add BlockBreakProgressUpdateEvent
 
+        ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(entityId, pos, progress); // Confetti - BroadcastPacketEvent - move packet creation out of loop
+        new com.vulpeslab.confetti.event.BroadcastPacketEvent(this.getWorld(), packet, entity == null ? null : entity.getBukkitEntity()).callEvent(); // Confetti - BroadcastPacketEvent
         while (iterator.hasNext()) {
             ServerPlayer entityplayer = (ServerPlayer) iterator.next();
 
@@ -1701,7 +1703,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
                 // CraftBukkit end
 
                 if (d0 * d0 + d1 * d1 + d2 * d2 < 1024.0D) {
-                    entityplayer.connection.send(new ClientboundBlockDestructionPacket(entityId, pos, progress));
+                    entityplayer.connection.send(packet); // Confetti - BroadcastPacketEvent - move packet creation out of loop
                 }
             }
         }
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 33b7f0f14c826e9baa8b562a852cd4d46882622c..9a9ef95e6bdba46e4befa4e1091323dcbf5386c2 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -72,6 +72,7 @@ import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.PlayerAdvancements;
 import net.minecraft.server.RegistryLayer;
 import net.minecraft.server.ServerScoreboard;
+import net.minecraft.server.level.ChunkMap;
 import net.minecraft.server.level.ClientInformation;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
@@ -1209,6 +1210,7 @@ public abstract class PlayerList {
     }
 
     public void broadcast(@Nullable net.minecraft.world.entity.player.Player player, double x, double y, double z, double distance, ResourceKey<Level> worldKey, Packet<?> packet) {
+        if (player != null) new com.vulpeslab.confetti.event.BroadcastPacketEvent(this.getServer().getLevel(worldKey).getWorld(), packet, player.getBukkitEntity()).callEvent(); // Confetti - BroadcastPacketEvent
         for (int i = 0; i < this.players.size(); ++i) {
             ServerPlayer entityplayer = (ServerPlayer) this.players.get(i);
 
