From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 31 Oct 2025 03:02:25 +0900
Subject: [PATCH] Only apply item gravity if not on ground


diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index ffc754aa6ed61f62a0c94e9117f3008d24c0c163..f249c19cc58057f69da0d3068be1abb765b7e037 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -175,7 +175,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
             } else if (this.isInLava() && this.getFluidHeight(FluidTags.LAVA) > 0.10000000149011612D) {
                 this.setUnderLavaMovement();
             } else {
-                this.applyGravity();
+                if (!this.onGround()) this.applyGravity(); // Confetti - only apply gravity if not on ground
             }
 
             if (this.level().isClientSide) {
