From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 31 Oct 2025 03:02:25 +0900
Subject: [PATCH] Only apply item gravity if not on ground


diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index 499646ad26d148953abf241d90e053870fedb708..c1ba0da44b40d75cbe76698c69810261397bcefb 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -177,7 +177,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
             } else if (this.isInLava() && this.getFluidHeight(FluidTags.LAVA) > 0.10000000149011612D) {
                 this.setUnderLavaMovement();
             } else {
-                this.applyGravity();
+                if (!this.onGround()) this.applyGravity(); // Confetti - only apply gravity if not on ground
             }
 
             if (this.level().isClientSide) {
